---
title: "Format Data for import into MARVEL"
output: html_document
date: "2023-12-21"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

The purpose of this script is to format expression data from seurat and SJ outputs from STARsolo for MARVEL. It is dependent on running the scripts 01 and 02. Please use docker image setbp1_alternative_splicing:1.0.5.

## Load Packages

```{r load packages}
suppressPackageStartupMessages({
  # Load Lasseigne Lab standard packages
  library(here)
  library(styler)
  library(lintr)

  # Load Seurat and other single-cell packages
  library(tidyverse)
  library(Seurat)
  library(patchwork)
  library(harmony)
  library(presto)

  # Load MARVEL package
  library(MARVEL)

  # Load adjunct MARVEL packages
  library(ggnewscale)
  library(ggrepel)
  library(reshape2)
  library(plyr)
  library(textclean)
  library(AnnotationDbi)
  library(clusterProfiler)
  library(org.Hs.eg.db)
  library(org.Mm.eg.db)
  library(gtools)
  library(GenomicRanges)
  library(IRanges)
  library(S4Vectors)
  library(wiggleplotr)
  library(Matrix)
  library(data.table)
  library(gridExtra)

  # Load this for data wrangling ease
  library(MatrixExtra)
  # Disable this option to use the following indexing method of checking values
  options("MatrixExtra.quick_show" = FALSE)
})

# set seed
set.seed(123)

# source functions
source(here::here("src", "marvel", "functions.R"))
```

## Import Expression Data

First, you need to import and data wrangle the Seurat expression data.

```{r import seurat expression data}
annotated_brain_samples <- readRDS(
  here::here("data", "seurat", "annotated_brain_samples.rds")
)

gene_norm <- annotated_brain_samples@assays[["RNA"]]@layers[["data"]]
gene_norm <- as(gene_norm, "TsparseMatrix")
gene_norm[gene_norm[, 1] != 0, ][1:5, 1:5]

gene_metadata <- annotated_brain_samples@meta.data
gene_metadata <- tibble::rownames_to_column(gene_metadata, "cell.id")
head(gene_metadata)

gene_features <- rownames(annotated_brain_samples)
head(gene_features)
gene_features <- as.data.frame(gene_features)
colnames(gene_features) <- "gene_short_name"

gene_counts <- annotated_brain_samples@assays[["RNA"]]@layers[["counts"]]
gene_counts <- as(gene_counts, "dgTMatrix")
gene_counts[gene_counts[, 1] != 0, ][1:5, 1:5]
```

These software developers have a separate argument for df.gene.feature/df.count.feature and df.gene.pheno/df.count.pheno but they are the same files!!! So I am calling them gene_features, and gene_metadata.

## Import Splice Junction Data

The cell IDs of splice junction counts also have to match the expression cell IDs, meaning I need to data wrangle these files together to make a single splice junction out file.

I made a function to extract splice junction information for each sample to simplify the data wrangling required for this project, so we can just use that.

```{r use function to extract splice information}
# sample J1
J1_sj_info <- split_sj_info(
  sample_id = "J1",
  condition = "het"
)
# sample J2
J2_sj_info <- split_sj_info(
  sample_id = "J2",
  condition = "ctrl"
)
# sample J3
J3_sj_info <- split_sj_info(
  sample_id = "J3",
  condition = "ctrl"
)
# sample J4
J4_sj_info <- split_sj_info(
  sample_id = "J4",
  condition = "ctrl"
)
# sample J13
J13_sj_info <- split_sj_info(
  sample_id = "J13",
  condition = "het"
)
# sample J15
J15_sj_info <- split_sj_info(
  sample_id = "J15",
  condition = "het"
)
```

Now massage these separate SJ files into one larger expression matrix. Data is too large to pivot longer so need to do this method of joining specifically. The next solution fills in values based on the indexed feature name on where they should be.

```{r massage data frames}
# Make list of all matrices - make sure these are in the order of your gene data
matrices <- list(
  J1_sj_info, J13_sj_info, J15_sj_info, J2_sj_info, J3_sj_info,
  J4_sj_info
)

# Identify all unique features
all_features <- unique(unlist(lapply(matrices, rownames)))

# Align and combine matrices using custom function
combined_matrix <- do.call(cbind, lapply(matrices, align_matrix, all_features))

# Add correct colnames
colnames(combined_matrix) <- gene_metadata$cell.id

# Coerce to correct format
combined_matrix <- as(combined_matrix, "TsparseMatrix")

# Check values where each new sample starts
combined_matrix[combined_matrix[, 1] != 0, ][1:5, 1:5]

J1_sj_info[J1_sj_info[, 1] != 0, ][1:5, 1:5]

# Check each sample - J13
combined_matrix[combined_matrix[, 5039] != 0, ][1:5, 5039:5047]

J13_sj_info[J13_sj_info[, 1] != 0, ][1:5, 1:9]

# Check each sample - J15
combined_matrix[combined_matrix[, 14474] != 0, ][1:5, 14474:14482]

J15_sj_info[J15_sj_info[, 1] != 0, ][1:5, 1:9]

# Check each sample - J2
combined_matrix[combined_matrix[, 23427] != 0, ][1:5, 23427:23435]

J2_sj_info[J2_sj_info[, 1] != 0, ][1:5, 1:9]

# Check each sample - J3
combined_matrix[combined_matrix[, 30507] != 0, ][1:5, 30507:30515]

J3_sj_info[J3_sj_info[, 1] != 0, ][1:5, 1:9]

# Check each sample - J4
combined_matrix[combined_matrix[, 40978] != 0, ][1:5, 40978:40986]

J4_sj_info[J4_sj_info[, 1] != 0, ][1:5, 1:9]

# rename matrix for ease
all_sj_counts <- combined_matrix

# get metadata (colnames)
sj_metadata <- colnames(combined_matrix)
sj_metadata <- as.data.frame(sj_metadata)
colnames(sj_metadata) <- "cell.id"

# get splice junction location names
sj_features <- rownames(combined_matrix)
sj_features <- as.data.frame(sj_features)
colnames(sj_features) <- "coord.intron"
```

### Get UMAP coordinates

```{r get UMAP coordinates}
umap_coords <-
  annotated_brain_samples@reductions[["umap_harmony"]]@cell.embeddings

umap_coords <- tibble::rownames_to_column(as.data.frame(umap_coords), "cell.id")

colnames(umap_coords) <- c("cell.id", "x", "y")
```

### Get GTF

```{r get gtf in r env}
m31_gtf <- as.data.frame(fread("/data/project/lasseigne_lab/GENOME_dir/GENCODE_mm39/release_M31/GTF/gencode.vM31.primary_assembly.annotation.gtf"), sep = "\t", header = FALSE, stringsAsFactors = FALSE)
# rename columns
colnames(m31_gtf) <- c("V1", "V2", "V3", "V4", "V5", "V6", "V7", "V8", "V9")
# drop chr from chromosome numbers in gtf column 1
m31_gtf$V1 <- substr(m31_gtf$V1, 4, nchar(m31_gtf$V1))
```

Now, I'm going to try making the marvel object.

## Make MARVEL object

```{r make marvel object}
setbp1_marvel <- CreateMarvelObject.10x(
  gene.norm.matrix = gene_norm,
  gene.norm.pheno = gene_metadata,
  gene.norm.feature = gene_features,
  gene.count.matrix = gene_counts,
  gene.count.pheno = gene_metadata,
  gene.count.feature = gene_features,
  sj.count.matrix = all_sj_counts,
  sj.count.pheno = sj_metadata,
  sj.count.feature = sj_features,
  pca = umap_coords,
  gtf = m31_gtf
)
```

## Pre-process MARVEL object

First, we annotate the gene metadata.

```{r annotate gene metadta}
setbp1_marvel <- AnnotateGenes.10x(MarvelObject = setbp1_marvel)

head(setbp1_marvel$gene.metadata)

table(setbp1_marvel$gene.metadata$gene_type)
```

Now, we annotate the splice junction metadata.

```{r annotate junction metadata}
setbp1_marvel <- AnnotateSJ.10x(MarvelObject = setbp1_marvel)

head(setbp1_marvel$sj.metadata)
```

Now, we want to validate these splice junctions whose start and end are mapped to the same gene (no multi-gene splicing is included downstream because of this step).

```{r validate splice junctions}
setbp1_marvel <- ValidateSJ.10x(MarvelObject = setbp1_marvel)
```

Lastly, we will remove gene that are non-coding genes.

```{r subset cds genes}
setbp1_marvel <- FilterGenes.10x(
  MarvelObject = setbp1_marvel,
  gene.type = "protein_coding"
)
```

Last, before proceeding, we must check the alignment of the marvel object.

```{r check alignment}
setbp1_marvel <- CheckAlignment.10x(MarvelObject = setbp1_marvel)
```
Expected output:

Matching gene list in normalised and raw (count) gene matrices...
21935 genes found in normalised gene matrix
56979 genes found in raw (count) gene matrix
21935 overlapping genes found and retained
Matching cells across normalised gene, raw (count) gene, raw (count) SJ matrices...
51465 cells found in normalised gene matrix
51465 cells found in raw (count) gene matrix
51465 cells found in raw (count) SJ matrix
51465 overlapping cells found and retained
Checking column (sample) alignment...
Sample metadata and normalised gene matrix column names MATCHED
Normalised and raw (count) gene matrix column names MATCHED
Raw (Count) gene and SJ matrix column names MATCHED
Checking row (gene names/SJ coordinates) alignment...
Gene metadata and normalised gene matrix row names MATCHED
Normalised and raw (count) gene matrix row names MATCHED
SJ metadata and raw (count) SJ matrix row names MATCHED
51465 cells and 15633 genes consisting of 186562 splice junctions included for further analysis

Looks good! Now we can save our marvel object before proceeding.

```{r save marvel object}
# save object
saveRDS(setbp1_marvel,
  file = here::here("data", "marvel", "setbp1_marvel_aligned.rds")
)
```

#### Style

```{r tidy script}
style_file("01_format_MARVEL_data.Rmd")

lint("01_format_MARVEL_data.Rmd",
  linters = linters_with_defaults(
    object_length_linter = NULL,
    object_name_linter = NULL,
    object_usage_linter = NULL
  )
)
```


#### Software versions

My software versions will be commented below.

```{r versions}
sessionInfo()
```
R version 4.3.1 (2023-06-16)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] MatrixExtra_0.1.14    gridExtra_2.3         Matrix_1.6-1.1       
 [4] wiggleplotr_1.24.0    GenomicRanges_1.52.1  GenomeInfoDb_1.36.4  
 [7] gtools_3.9.4          org.Mm.eg.db_3.17.0   org.Hs.eg.db_3.17.0  
[10] clusterProfiler_4.8.3 AnnotationDbi_1.62.2  IRanges_2.34.1       
[13] S4Vectors_0.38.2      Biobase_2.60.0        BiocGenerics_0.46.0  
[16] textclean_0.9.3       plyr_1.8.9            reshape2_1.4.4       
[19] ggrepel_0.9.4         ggnewscale_0.4.9      MARVEL_2.0.5         
[22] presto_1.0.0          data.table_1.14.8     harmony_1.1.0        
[25] Rcpp_1.0.11           patchwork_1.1.3       Seurat_5.0.0         
[28] SeuratObject_5.0.0    sp_2.1-1              lubridate_1.9.3      
[31] forcats_1.0.0         stringr_1.5.1         dplyr_1.1.3          
[34] purrr_1.0.2           readr_2.1.4           tidyr_1.3.0          
[37] tibble_3.2.1          ggplot2_3.4.4         tidyverse_2.0.0      
[40] lintr_3.1.0           styler_1.10.2         here_1.0.1           

loaded via a namespace (and not attached):
  [1] fs_1.6.3                matrixStats_1.0.0       spatstat.sparse_3.0-3  
  [4] bitops_1.0-7            enrichplot_1.20.3       HDO.db_0.99.1          
  [7] httr_1.4.7              RColorBrewer_1.1-3      tools_4.3.1            
 [10] sctransform_0.4.1       backports_1.4.1         utf8_1.2.4             
 [13] R6_2.5.1                lazyeval_0.2.2          uwot_0.1.16            
 [16] withr_2.5.2             progressr_0.14.0        cli_3.6.1              
 [19] spatstat.explore_3.2-5  fastDummies_1.7.3       scatterpie_0.2.1       
 [22] spatstat.data_3.0-3     ggridges_0.5.4          pbapply_1.7-2          
 [25] yulab.utils_0.1.0       gson_0.1.0              DOSE_3.26.2            
 [28] R.utils_2.12.2          parallelly_1.36.0       rstudioapi_0.15.0      
 [31] RSQLite_2.3.1           generics_0.1.3          gridGraphics_0.5-1     
 [34] ica_1.0-3               spatstat.random_3.2-1   GO.db_3.17.0           
 [37] fansi_1.0.5             abind_1.4-5             R.methodsS3_1.8.2      
 [40] lifecycle_1.0.4         yaml_2.3.7              qvalue_2.32.0          
 [43] Rtsne_0.16              grid_4.3.1              blob_1.2.4             
 [46] promises_1.2.1          crayon_1.5.2            miniUI_0.1.1.1         
 [49] lattice_0.21-8          cowplot_1.1.1           KEGGREST_1.40.1        
 [52] pillar_1.9.0            knitr_1.44              fgsea_1.26.0           
 [55] future.apply_1.11.0     codetools_0.2-19        fastmatch_1.1-4        
 [58] leiden_0.4.3            glue_1.6.2              downloader_0.4         
 [61] ggfun_0.1.3             rex_1.2.1               remotes_2.4.2.1        
 [64] float_0.3-2             treeio_1.24.3           vctrs_0.6.4            
 [67] png_0.1-8               spam_2.10-0             gtable_0.3.4           
 [70] cachem_1.0.8            xfun_0.40               mime_0.12              
 [73] tidygraph_1.2.3         survival_3.5-7          ellipsis_0.3.2         
 [76] fitdistrplus_1.1-11     ROCR_1.0-11             nlme_3.1-163           
 [79] ggtree_3.8.2            bit64_4.0.5             RcppAnnoy_0.0.21       
 [82] rprojroot_2.0.3         R.cache_0.16.0          irlba_2.3.5.1          
 [85] KernSmooth_2.23-22      colorspace_2.1-0        DBI_1.1.3              
 [88] tidyselect_1.2.0        processx_3.8.2          bit_4.0.5              
 [91] compiler_4.3.1          xml2_1.3.5              desc_1.4.2             
 [94] plotly_4.10.3           shadowtext_0.1.2        scales_1.2.1           
 [97] lmtest_0.9-40           callr_3.7.3             digest_0.6.33          
[100] goftest_1.2-3           spatstat.utils_3.0-4    rmarkdown_2.25         
[103] RhpcBLASctl_0.23-42     XVector_0.40.0          htmltools_0.5.6.1      
[106] pkgconfig_2.0.3         fastmap_1.1.1           rlang_1.1.2            
[109] htmlwidgets_1.6.2       shiny_1.7.5.1           farver_2.1.1           
[112] zoo_1.8-12              jsonlite_1.8.7          BiocParallel_1.34.2    
[115] qdapRegex_0.7.8         GOSemSim_2.26.1         R.oo_1.25.0            
[118] RCurl_1.98-1.12         magrittr_2.0.3          GenomeInfoDbData_1.2.10
[121] ggplotify_0.1.2         dotCall64_1.1-0         munsell_0.5.0          
[124] ape_5.7-1               viridis_0.6.4           reticulate_1.34.0      
[127] stringi_1.8.1           ggraph_2.1.0            zlibbioc_1.46.0        
[130] MASS_7.3-60             parallel_4.3.1          listenv_0.9.0          
[133] deldir_1.0-9            Biostrings_2.68.1       graphlayouts_1.0.1     
[136] splines_4.3.1           tensor_1.5              hms_1.1.3              
[139] ps_1.7.5                igraph_1.5.1            spatstat.geom_3.2-7    
[142] RcppHNSW_0.5.0          evaluate_0.22           tzdb_0.4.0             
[145] tweenr_2.0.2            httpuv_1.6.12           RANN_2.6.1             
[148] polyclip_1.10-6         future_1.33.0           scattermore_1.2        
[151] ggforce_0.4.1           xtable_1.8-4            RSpectra_0.16-1        
[154] tidytree_0.4.5          cyclocomp_1.1.1         later_1.3.1            
[157] viridisLite_0.4.2       aplot_0.2.2             memoise_2.0.1          
[160] cluster_2.1.4           timechange_0.2.0        globals_0.16.2         
[163] xmlparsedata_1.0.5 