---
title: "MARVEL Differential Analysis"
author: "Emma Jones"
date: "2024-01-08"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache.lazy = FALSE)
```

The purpose of this script is to run differential analysis with MARVEL. It is dependent on running all seurat scripts, and marvel script 01. Please use docker image setbp1_alternative_splicing:1.0.5.

## Load Packages

```{r load packages}
suppressPackageStartupMessages({
  # Load Lasseigne Lab standard packages
  library(here)
  library(styler)
  library(lintr)
  library(tidyverse)
  library(patchwork)

  # Load MARVEL package
  library(MARVEL)

  # Load adjunct MARVEL packages
  library(ggnewscale)
  library(ggrepel)
  library(reshape2)
  library(plyr)
  library(stringr)
  library(textclean)
  library(AnnotationDbi)
  library(clusterProfiler)
  library(org.Hs.eg.db)
  library(org.Mm.eg.db)
  library(gtools)
  library(GenomicRanges)
  library(IRanges)
  library(S4Vectors)
  library(wiggleplotr)
  library(Matrix)
  library(data.table)
  library(gridExtra)
})

# set seed
set.seed(123)

# source functions
source(here::here("src", "marvel", "functions.R"))
```

## Import MARVEL Object

```{r import marvel object}
setbp1_marvel <- read_rds(here::here(
  "data", "marvel",
  "setbp1_marvel_aligned.rds"
))
# Retrieve sample metadata
sample_metadata <- setbp1_marvel$sample.metadata
```

## Explore expression

We do this in a pairwise fashion, and need to determine which expression thresholds to use. This will likely change for the comparisons being done.

```{r define cell groups}
# Group 1 (reference)
index <- which(sample_metadata$cell_type == "Excitatory Neurons" &
  sample_metadata$seq_folder == "mutant")
cell.ids.1 <- sample_metadata[index, "cell.id"]
length(cell.ids.1)

# Group 2
index <- which(sample_metadata$cell_type == "Excitatory Neurons" &
  sample_metadata$seq_folder == "wildtype")
cell.ids.2 <- sample_metadata[index, "cell.id"]
length(cell.ids.2)
```

Now, we know that on average, neurons express more genes than glia, so only using neurons for expression cutoffs means that the cutoffs should only be used for neuron-specific analyses.

```{r explore percent of cell expressing genes}
# Explore % of cells expressing genes
setbp1_marvel <- PlotPctExprCells.Genes.10x(
  MarvelObject = setbp1_marvel,
  cell.group.g1 = cell.ids.1,
  cell.group.g2 = cell.ids.2,
  min.pct.cells = 5
)

setbp1_marvel$pct.cells.expr$Gene$Plot

head(setbp1_marvel$pct.cells.expr$Gene$Data)
```

We can also look at this at the splice junction level.

```{r explore percent of cells expressing splice junctions}
setbp1_marvel <- PlotPctExprCells.SJ.10x(
  MarvelObject = setbp1_marvel,
  cell.group.g1 = cell.ids.1,
  cell.group.g2 = cell.ids.2,
  min.pct.cells.genes = 5,
  min.pct.cells.sj = 5,
  downsample = TRUE,
  downsample.pct.sj = 10
)

setbp1_marvel$pct.cells.expr$SJ$Plot

head(setbp1_marvel$pct.cells.expr$SJ$Data)
```

For, now I will use the default expression cutoffs for all differential analysis.

### Use in-house function for running marvel on a single cell type

```{r run marvel differential analysis on all cell types}
excitatory_marvel <- run_marvel_cell_type(
  marvel_object = setbp1_marvel,
  cell_type = "Excitatory Neurons"
)


inhibitory_marvel <- run_marvel_cell_type(
  marvel_object = setbp1_marvel,
  cell_type = "Inhibitory Neurons"
)


astrocytes_marvel <- run_marvel_cell_type(
  marvel_object = setbp1_marvel,
  cell_type = "Astrocytes"
)


oligodendrocytes_marvel <- run_marvel_cell_type(
  marvel_object = setbp1_marvel,
  cell_type = "Oligodendrocytes"
)


opcs_marvel <- run_marvel_cell_type(
  marvel_object = setbp1_marvel,
  cell_type = "OPCs"
)

microglia_marvel <- run_marvel_cell_type(
  marvel_object = setbp1_marvel,
  cell_type = "Microglia"
)


vascular_marvel <- run_marvel_cell_type(
  marvel_object = setbp1_marvel,
  cell_type = "Vascular Cells"
)
```

This next chunk of code is basically replicating what the function did, but with different filtering for the cell IDs Here, we are doing differential gene expression and splicing across all cell types, instead of within a single cell type.

```{r do for all cell types}
all_types_marvel <- setbp1_marvel

# Group 1 (reference)
index_1 <- which(sample_metadata$seq_folder == "mutant")
cell_ids_1 <- sample_metadata[index_1, "cell.id"]

# Group 2
index_2 <- which(sample_metadata$seq_folder == "wildtype")
cell_ids_2 <- sample_metadata[index_2, "cell.id"]

# Explore % of cells expressing genes
all_types_marvel <- PlotPctExprCells.Genes.10x(
  MarvelObject = all_types_marvel,
  cell.group.g1 = cell_ids_1,
  cell.group.g2 = cell_ids_2,
  min.pct.cells = 5
)

# Explore % of cells expressing junctions
all_types_marvel <- PlotPctExprCells.SJ.10x(
  MarvelObject = all_types_marvel,
  cell.group.g1 = cell_ids_1,
  cell.group.g2 = cell_ids_2,
  min.pct.cells.genes = 5,
  min.pct.cells.sj = 5,
  downsample = TRUE,
  downsample.pct.sj = 10
)

# Differential Splicing Analysis
all_types_marvel <- CompareValues.SJ.10x(
  MarvelObject = all_types_marvel,
  cell.group.g1 = cell_ids_1,
  cell.group.g2 = cell_ids_2,
  min.pct.cells.genes = 5,
  min.pct.cells.sj = 5,
  min.gene.norm = 1,
  seed = 1,
  n.iterations = 100,
  downsample = TRUE,
  show.progress = TRUE
)

# Differential Gene Analysis
all_types_marvel <- CompareValues.Genes.10x(
  MarvelObject = all_types_marvel,
  show.progress = TRUE
)

# Make volcano plot
all_types_marvel <- PlotDEValues.SJ.10x(
  MarvelObject = all_types_marvel,
  pval = 0.05,
  delta = 1,
  min.gene.norm = 1,
  anno = FALSE
)
# Assign kinds of iso-switching
all_types_marvel <- IsoSwitch.10x(
  MarvelObject = all_types_marvel,
  pval.sj = 0.05,
  delta.sj = 1,
  min.gene.norm = 1,
  pval.adj.gene = 0.05,
  log2fc.gene = 0.5
)

# Pull significant genes
significant_genes <- all_types_marvel[["SJ.Gene.Cor"]][["Data"]]$gene_short_name

# Save final
saveRDS(all_types_marvel,
  file = here::here(
    "data", "marvel",
    paste0("All_Cell_Types_marvel_object.rds")
  )
)
```


### Gene-splicing dynamics visalization - excitatory neurons

```{r what are the significant genes for this analysis}
# Pull significant genes
excitatory_sig_table <-
  excitatory_marvel[["SJ.Gene.Cor"]][["Data"]]

# Pull significant genes
excitatory_sig_genes <-
  excitatory_marvel[["SJ.Gene.Cor"]][["Data"]]$gene_short_name
```

```{r split into groups if needed}
# Group 1 (reference)
index_1 <- which(sample_metadata$seq_folder == "mutant")
cell_ids_1 <- sample_metadata[index_1, "cell.id"]

# Group 2
index_2 <- which(sample_metadata$seq_folder == "wildtype")
cell_ids_2 <- sample_metadata[index_2, "cell.id"]

# Save into list
cell_group_list <- list(
  "Mutant" = cell_ids_1,
  "Wildtype" = cell_ids_2
)
```

```{r visualize on gene and splice junction expression on umap}
# Plot cell groups
excitatory_marvel <- PlotValues.PCA.CellGroup.10x(
  MarvelObject = excitatory_marvel,
  cell.group.list = cell_group_list,
  legendtitle = "Cell group",
  type = "umap"
)

plot_group <- excitatory_marvel$adhocPlot$PCA$CellGroup

# Plot gene expression
excitatory_marvel <- PlotValues.PCA.Gene.10x(
  MarvelObject = excitatory_marvel,
  gene_short_name = "Kif21a",
  color.gradient = c("grey", "cyan", "green", "yellow", "red"),
  type = "umap"
)


plot_gene <- excitatory_marvel$adhocPlot$PCA$Gene

# Plot PSI
excitatory_marvel <- PlotValues.PCA.PSI.10x(
  MarvelObject = excitatory_marvel,
  coord.intron = "chr15:90881929:90882035",
  min.gene.count = 3,
  log2.transform = FALSE,
  color.gradient = c("grey", "cyan", "green", "yellow", "red"),
  type = "umap"
)

plot_sj <- excitatory_marvel$adhocPlot$PCA$PSI

# Arrange and view plots
grid.arrange(plot_group, plot_gene, plot_sj, nrow = 1)

# Save UMAP
png(here::here("results", "marvel_outputs", "excitatory_Kif21a.png"),
  width = 1200, height = 600
)
grid.arrange(plot_group, plot_gene, plot_sj, nrow = 1)
dev.off()
```


```{r test umap plotting function}
excitatory_marvel <- plot_marvel_umap(
  marvel_object = excitatory_marvel,
  prefix = "excitatory",
  gene = "Kif21a",
  sj_loc = "chr15:90881929:90882035"
)
```

## Gene Ontology Analysis

Unfortunately, at least 10 genes are required for GO analysis, so I am skipping this.

## Candidate Gene Analysis

It looks like I can make as many cell groups and label as I like. This may help us for comparing expression of a gene or splice junction across all cell types for both mutant and control.

### Plot SJ position

```{r plot SJ position}
#install.packages(here::here("bin", "custom_packages", "wiggleplotr_1.18.0.tar.gz"), repos = NULL, type = "source")

excitatory_marvel <- adhocGene_PlotSJPosition_EJ(
  MarvelObject = excitatory_marvel,
  coord.intron = "chr15:90881929:90882035",
  rescale_introns = FALSE,
  show.protein.coding.only = TRUE,
  anno.label.size = 1.5
)

excitatory_marvel$adhocGene$SJPosition$Plot
```

#### Style

```{r tidy script}
style_file("02_MARVEL_differential_analysis.Rmd")

lint("02_MARVEL_differential_analysis.Rmd",
  linters = linters_with_defaults(
    object_length_linter = NULL,
    object_name_linter = NULL,
    object_usage_linter = NULL
  )
)
```


#### Software versions

My software versions will be commented below.

```{r versions}
sessionInfo()
```
