---
title: "MARVEL Differential Analysis"
author: "Emma Jones"
date: "2024-01-08"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

The purpose of this script is to run differential analysis with MARVEL. It is dependent on running all seurat scripts, and marvel script 01. Please use docker image setbp1_alternative_splicing:1.0.5.

## Load Packages

```{r load packages}
suppressPackageStartupMessages({
  # Load Lasseigne Lab standard packages
  library(here)
  library(styler)
  library(lintr)
  library(tidyverse)
  library(patchwork)

  # Load MARVEL package
  library(MARVEL)

  # Load adjunct MARVEL packages
  library(ggnewscale)
  library(ggrepel)
  library(reshape2)
  library(plyr)
  library(stringr)
  library(textclean)
  library(AnnotationDbi)
  library(clusterProfiler)
  library(org.Hs.eg.db)
  library(org.Mm.eg.db)
  library(gtools)
  library(GenomicRanges)
  library(IRanges)
  library(S4Vectors)
  library(wiggleplotr)
  library(Matrix)
  library(data.table)
  library(gridExtra)
})

# set seed
set.seed(123)

# source functions
source(here::here("src", "marvel", "functions.R"))
```

## Import MARVEL Object

```{r import marvel object}
setbp1_marvel <- readRDS(here::here(
  "data", "marvel",
  "setbp1_marvel_aligned.rds"
))
```

## Explore expression

We do this in a pairwise fashion, and need to determine which expression thresholds to use. This will likely change for the comparisons being done. For now, I'll compare excitatory to inhibitory neurons since those have the biggest numbers of cells.

```{r define cell groups}
# Retrieve sample metadata
sample_metadata <- setbp1_marvel$sample.metadata

# Group 1 (reference)
index <- which(sample_metadata$cell_type == "Excitatory Neurons")
cell.ids.1 <- sample_metadata[index, "cell.id"]
length(cell.ids.1)

# Group 2
index <- which(sample_metadata$cell_type == "Inhibitory Neurons")
cell.ids.2 <- sample_metadata[index, "cell.id"]
length(cell.ids.2)
```

Now, we know that on average, neurons express more genes than glia, so only using neurons for expression cutoffs means that the cutoffs should only be used for neuron-specific analyses.

```{r explore percent of cell expressing genes}
# Explore % of cells expressing genes
setbp1_marvel <- PlotPctExprCells.Genes.10x(
  MarvelObject = setbp1_marvel,
  cell.group.g1 = cell.ids.1,
  cell.group.g2 = cell.ids.2,
  min.pct.cells = 5
)

setbp1_marvel$pct.cells.expr$Gene$Plot

head(setbp1_marvel$pct.cells.expr$Gene$Data)
```

We can also look at this at the splice junction level.

```{r explore percent of cells expressing splice junctions}
setbp1_marvel <- PlotPctExprCells.SJ.10x(
  MarvelObject = setbp1_marvel,
  cell.group.g1 = cell.ids.1,
  cell.group.g2 = cell.ids.2,
  min.pct.cells.genes = 5,
  min.pct.cells.sj = 5,
  downsample = TRUE,
  downsample.pct.sj = 10
)

setbp1_marvel$pct.cells.expr$SJ$Plot

head(setbp1_marvel$pct.cells.expr$SJ$Data)
```

For, now I will use the default expression cutoffs for differential analysis.

## Differential Analysis

First we can do differential splicing analysis

```{r differential splicing analysis}
setbp1_marvel <- CompareValues.SJ.10x(
  MarvelObject = setbp1_marvel,
  cell.group.g1 = cell.ids.1,
  cell.group.g2 = cell.ids.2,
  min.pct.cells.genes = 10,
  min.pct.cells.sj = 10,
  min.gene.norm = 1.0,
  seed = 1,
  n.iterations = 100,
  downsample = TRUE,
  show.progress = TRUE
)
```

Then, we can do differential gene analysis

```{r differential gene analysis}
setbp1_marvel <- CompareValues.Genes.10x(
  MarvelObject = setbp1_marvel,
  show.progress = TRUE
)

head(setbp1_marvel$DE$SJ$Table)
```
### Volcano plots!

```{r make splicing volcano plot}
setbp1_marvel <- PlotDEValues.SJ.10x(
  MarvelObject = setbp1_marvel,
  pval = 0.1,
  delta = 1,
  min.gene.norm = 0.5,
  anno = FALSE
)

setbp1_marvel$DE$SJ$VolcanoPlot$SJ$Plot
```

This volcano plot is pretty wack, but maybe we can work on this a bit. For now, I am going to work with the whole vignette with my example data.

```{r assign dynamics}
setbp1_marvel <- IsoSwitch.10x(
  MarvelObject = setbp1_marvel,
  pval.sj = 0.05,
  delta.sj = 1,
  min.gene.norm = 1.0,
  pval.adj.gene = 0.05,
  log2fc.gene = 0.5
)

setbp1_marvel$SJ.Gene.Cor$Proportion$Plot
```
### Gene-splicing dynamics

```{r visualize gene-splicing dynamics}
# Save into list
cell.group.list <- list(
  "Exitatory" = cell.ids.1,
  "Inhibitory" = cell.ids.2
)

# Plot cell groups
setbp1_marvel <- PlotValues.PCA.CellGroup.10x(
  MarvelObject = setbp1_marvel,
  cell.group.list = cell.group.list,
  legendtitle = "Cell group",
  type = "umap"
)

plot_group <- setbp1_marvel$adhocPlot$PCA$CellGroup

# Plot gene expression
setbp1_marvel <- PlotValues.PCA.Gene.10x(
  MarvelObject = setbp1_marvel,
  gene_short_name = "Dst",
  color.gradient = c("grey", "cyan", "green", "yellow", "red"),
  type = "umap"
)


plot_gene <- setbp1_marvel$adhocPlot$PCA$Gene

# Plot PSI
setbp1_marvel <- PlotValues.PCA.PSI.10x(
  MarvelObject = setbp1_marvel,
  coord.intron = "chr1:34007985:34045271",
  min.gene.count = 3,
  log2.transform = FALSE,
  color.gradient = c("grey", "cyan", "green", "yellow", "red"),
  type = "umap"
)

plot_sj <- setbp1_marvel$adhocPlot$PCA$PSI

# Arrange and view plots
grid.arrange(plot_group, plot_gene, plot_sj, nrow = 1)
```

#### Style

```{r tidy script}
style_file("02_MARVEL_differential_analysis.Rmd")

lint("02_MARVEL_differential_analysis.Rmd",
  linters = linters_with_defaults(
    object_length_linter = NULL,
    object_name_linter = NULL,
    object_usage_linter = NULL
  )
)
```


#### Software versions

My software versions will be commented below.

```{r versions}
sessionInfo()
```
