---
title: "MARVEL Differential Analysis"
author: "Emma Jones"
date: "2024-01-08"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache.lazy = FALSE)
```

The purpose of this script is to run differential analysis with MARVEL. It is dependent on running all seurat scripts, and marvel script 01. Please use docker image setbp1_alternative_splicing:1.0.5.

## Load Packages

```{r load packages}
suppressPackageStartupMessages({
  # Load Lasseigne Lab standard packages
  library(here)
  library(styler)
  library(lintr)
  library(tidyverse)
  library(patchwork)

  # Load MARVEL package
  library(MARVEL)

  # Load adjunct MARVEL packages
  library(ggnewscale)
  library(ggrepel)
  library(reshape2)
  library(plyr)
  library(stringr)
  library(textclean)
  library(AnnotationDbi)
  library(clusterProfiler)
  library(org.Hs.eg.db)
  library(org.Mm.eg.db)
  library(gtools)
  library(GenomicRanges)
  library(IRanges)
  library(S4Vectors)
  library(wiggleplotr)
  library(Matrix)
  library(data.table)
  library(gridExtra)
})

# set seed
set.seed(123)

# source functions
source(here::here("src", "marvel", "functions.R"))
```

## Import MARVEL Object

```{r import marvel object}
setbp1_marvel <- read_rds(here::here(
  "data", "marvel",
  "setbp1_marvel_aligned.rds"
))
# Retrieve sample metadata
sample_metadata <- setbp1_marvel$sample.metadata
```

## Explore expression

We do this in a pairwise fashion, and need to determine which expression thresholds to use. This will likely change for the comparisons being done.

```{r define cell groups for excitatory neurons}
# Group 1 (reference)
index <- which(sample_metadata$cell_type == "Excitatory Neurons" &
  sample_metadata$seq_folder == "mutant")
cell_ids_1 <- sample_metadata[index, "cell.id"]
length(cell_ids_1)

# Group 2
index <- which(sample_metadata$cell_type == "Excitatory Neurons" &
  sample_metadata$seq_folder == "wildtype")
cell_ids_2 <- sample_metadata[index, "cell.id"]
length(cell_ids_2)

```

Now, we know that on average, neurons express more genes than glia, so only using neurons for expression cutoffs means that the cutoffs should only be used for neuron-specific analyses.

```{r explore percent of cell expressing genes}
# Explore % of cells expressing genes
setbp1_marvel <- PlotPctExprCells.Genes.10x(
  MarvelObject = setbp1_marvel,
  cell.group.g1 = cell_ids_1,
  cell.group.g2 = cell_ids_2,
  min.pct.cells = 5
)

setbp1_marvel$pct.cells.expr$Gene$Plot

head(setbp1_marvel$pct.cells.expr$Gene$Data)
```

We can also look at this at the splice junction level.

```{r explore percent of cells expressing splice junctions}
setbp1_marvel <- PlotPctExprCells.SJ.10x(
  MarvelObject = setbp1_marvel,
  cell.group.g1 = cell_ids_1,
  cell.group.g2 = cell_ids_2,
  min.pct.cells.genes = 5,
  min.pct.cells.sj = 5,
  downsample = TRUE,
  downsample.pct.sj = 10
)

setbp1_marvel$pct.cells.expr$SJ$Plot

head(setbp1_marvel$pct.cells.expr$SJ$Data)
```

For, now I will use the default expression cutoffs for all differential analysis.

### Use in-house function for running marvel on a single cell type

```{r run marvel differential analysis on all cell types}
excitatory_marvel <- run_marvel_cell_type(
  marvel_object = setbp1_marvel,
  cell_type = "Excitatory Neurons"
)


inhibitory_marvel <- run_marvel_cell_type(
  marvel_object = setbp1_marvel,
  cell_type = "Inhibitory Neurons"
)


astrocytes_marvel <- run_marvel_cell_type(
  marvel_object = setbp1_marvel,
  cell_type = "Astrocytes"
)


oligodendrocytes_marvel <- run_marvel_cell_type(
  marvel_object = setbp1_marvel,
  cell_type = "Oligodendrocytes"
)


opcs_marvel <- run_marvel_cell_type(
  marvel_object = setbp1_marvel,
  cell_type = "OPCs"
)

microglia_marvel <- run_marvel_cell_type(
  marvel_object = setbp1_marvel,
  cell_type = "Microglia"
)


vascular_marvel <- run_marvel_cell_type(
  marvel_object = setbp1_marvel,
  cell_type = "Vascular Cells"
)
```

This next chunk of code is basically replicating what the function did, but with different filtering for the cell IDs Here, we are doing differential gene expression and splicing across all cell types, instead of within a single cell type.

```{r do for all cell types}
all_types_marvel <- setbp1_marvel

# Group 1 (reference)
index_1 <- which(sample_metadata$seq_folder == "mutant")
cell_ids_1 <- sample_metadata[index_1, "cell.id"]

# Group 2
index_2 <- which(sample_metadata$seq_folder == "wildtype")
cell_ids_2 <- sample_metadata[index_2, "cell.id"]

# Explore % of cells expressing genes
all_types_marvel <- PlotPctExprCells.Genes.10x(
  MarvelObject = all_types_marvel,
  cell.group.g1 = cell_ids_1,
  cell.group.g2 = cell_ids_2,
  min.pct.cells = 5
)

# Explore % of cells expressing junctions
all_types_marvel <- PlotPctExprCells.SJ.10x(
  MarvelObject = all_types_marvel,
  cell.group.g1 = cell_ids_1,
  cell.group.g2 = cell_ids_2,
  min.pct.cells.genes = 5,
  min.pct.cells.sj = 5,
  downsample = TRUE,
  downsample.pct.sj = 10
)

# Differential Splicing Analysis
all_types_marvel <- CompareValues.SJ.10x(
  MarvelObject = all_types_marvel,
  cell.group.g1 = cell_ids_1,
  cell.group.g2 = cell_ids_2,
  min.pct.cells.genes = 5,
  min.pct.cells.sj = 5,
  min.gene.norm = 1,
  seed = 1,
  n.iterations = 100,
  downsample = TRUE,
  show.progress = TRUE
)

# Differential Gene Analysis
all_types_marvel <- CompareValues.Genes.10x(
  MarvelObject = all_types_marvel,
  show.progress = TRUE
)

# Make volcano plot
all_types_marvel <- PlotDEValues.SJ.10x(
  MarvelObject = all_types_marvel,
  pval = 0.05,
  delta = 1,
  min.gene.norm = 1,
  anno = FALSE
)
# Assign kinds of iso-switching
all_types_marvel <- IsoSwitch.10x(
  MarvelObject = all_types_marvel,
  pval.sj = 0.05,
  delta.sj = 1,
  min.gene.norm = 1,
  pval.adj.gene = 0.05,
  log2fc.gene = 0.5
)

# Pull significant genes
significant_genes <- all_types_marvel[["SJ.Gene.Cor"]][["Data"]]$gene_short_name

# Save final
saveRDS(all_types_marvel,
  file = here::here(
    "data", "marvel",
    paste0("All_Cell_Types_marvel_object.rds")
  )
)
```

#### Style

```{r tidy script}
style_file("02_MARVEL_differential_analysis.Rmd")

lint("02_MARVEL_differential_analysis.Rmd",
  linters = linters_with_defaults(
    object_length_linter = NULL,
    object_name_linter = NULL,
    object_usage_linter = NULL
  )
)
```


#### Software versions

My software versions will be commented below.

```{r versions}
sessionInfo()
```
R version 4.3.1 (2023-06-16)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] gridExtra_2.3         data.table_1.14.8     Matrix_1.6-1.1       
 [4] wiggleplotr_1.24.0    GenomicRanges_1.52.1  GenomeInfoDb_1.36.4  
 [7] gtools_3.9.4          org.Mm.eg.db_3.17.0   org.Hs.eg.db_3.17.0  
[10] clusterProfiler_4.8.3 AnnotationDbi_1.62.2  IRanges_2.34.1       
[13] S4Vectors_0.38.2      Biobase_2.60.0        BiocGenerics_0.46.0  
[16] textclean_0.9.3       plyr_1.8.9            reshape2_1.4.4       
[19] ggrepel_0.9.4         ggnewscale_0.4.9      MARVEL_2.0.5         
[22] patchwork_1.1.3       lubridate_1.9.3       forcats_1.0.0        
[25] stringr_1.5.1         dplyr_1.1.3           purrr_1.0.2          
[28] readr_2.1.4           tidyr_1.3.0           tibble_3.2.1         
[31] ggplot2_3.4.4         tidyverse_2.0.0       lintr_3.1.0          
[34] styler_1.10.2         here_1.0.1           

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3      jsonlite_1.8.7          rstudioapi_0.15.0      
  [4] magrittr_2.0.3          farver_2.1.1            rmarkdown_2.25         
  [7] fs_1.6.3                zlibbioc_1.46.0         vctrs_0.6.4            
 [10] memoise_2.0.1           RCurl_1.98-1.12         ggtree_3.8.2           
 [13] htmltools_0.5.6.1       qdapRegex_0.7.8         gridGraphics_0.5-1     
 [16] desc_1.4.2              cachem_1.0.8            igraph_1.5.1           
 [19] lifecycle_1.0.4         pkgconfig_2.0.3         gson_0.1.0             
 [22] R6_2.5.1                fastmap_1.1.1           GenomeInfoDbData_1.2.10
 [25] digest_0.6.33           aplot_0.2.2             enrichplot_1.20.3      
 [28] colorspace_2.1-0        ps_1.7.5                rprojroot_2.0.3        
 [31] RSQLite_2.3.1           labeling_0.4.3          fansi_1.0.5            
 [34] timechange_0.2.0        httr_1.4.7              polyclip_1.10-6        
 [37] compiler_4.3.1          remotes_2.4.2.1         bit64_4.0.5            
 [40] withr_2.5.2             downloader_0.4          backports_1.4.1        
 [43] BiocParallel_1.34.2     viridis_0.6.4           DBI_1.1.3              
 [46] ggforce_0.4.1           R.utils_2.12.2          MASS_7.3-60            
 [49] HDO.db_0.99.1           tools_4.3.1             scatterpie_0.2.1       
 [52] ape_5.7-1               R.oo_1.25.0             glue_1.6.2             
 [55] callr_3.7.3             nlme_3.1-163            R.cache_0.16.0         
 [58] GOSemSim_2.26.1         shadowtext_0.1.2        grid_4.3.1             
 [61] fgsea_1.26.0            generics_0.1.3          gtable_0.3.4           
 [64] tzdb_0.4.0              R.methodsS3_1.8.2       hms_1.1.3              
 [67] tidygraph_1.2.3         xml2_1.3.5              utf8_1.2.4             
 [70] XVector_0.40.0          pillar_1.9.0            yulab.utils_0.1.0      
 [73] splines_4.3.1           tweenr_2.0.2            treeio_1.24.3          
 [76] lattice_0.21-8          bit_4.0.5               tidyselect_1.2.0       
 [79] GO.db_3.17.0            Biostrings_2.68.1       knitr_1.44             
 [82] xfun_0.40               graphlayouts_1.0.1      rex_1.2.1              
 [85] stringi_1.8.1           lazyeval_0.2.2          ggfun_0.1.3            
 [88] yaml_2.3.7              evaluate_0.22           codetools_0.2-19       
 [91] ggraph_2.1.0            qvalue_2.32.0           ggplotify_0.1.2        
 [94] cli_3.6.1               munsell_0.5.0           processx_3.8.2         
 [97] Rcpp_1.0.11             png_0.1-8               parallel_4.3.1         
[100] blob_1.2.4              DOSE_3.26.2             bitops_1.0-7           
[103] tidytree_0.4.5          viridisLite_0.4.2       cyclocomp_1.1.1        
[106] scales_1.2.1            crayon_1.5.2            rlang_1.1.2            
[109] cowplot_1.1.1           fastmatch_1.1-4         KEGGREST_1.40.1  