---
title: "Analyze DE Gene and SJ by Cell type"
output: html_document
date: "2024-01-18"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

The purpose of this script is to analyze differential analysis results MARVEL. It is dependent on running all seurat scripts, and marvel scripts 01 and 02. Please use docker image setbp1_alternative_splicing:1.0.6, since you need ComplexHeatmap so make the UpSet plots.

## Load Packages

```{r load packages}
suppressPackageStartupMessages({
  # Load Lasseigne Lab standard packages
  library(here)
  library(styler)
  library(lintr)
  library(tidyverse)
  library(patchwork)
  library(ComplexHeatmap)
  library(viridis)

  # Load MARVEL package
  library(MARVEL)

  # Load adjunct MARVEL packages
  library(ggnewscale)
  library(ggrepel)
  library(reshape2)
  library(plyr)
  library(stringr)
  library(textclean)
  library(AnnotationDbi)
  library(clusterProfiler)
  library(org.Hs.eg.db)
  library(org.Mm.eg.db)
  library(gtools)
  library(GenomicRanges)
  library(IRanges)
  library(S4Vectors)
  library(wiggleplotr)
  library(Matrix)
  library(data.table)
  library(gridExtra)
})

# set seed
set.seed(123)

# source functions
source(here::here("src", "marvel", "functions.R"))
source(here::here("src", "marvel", "PlotSJPosition_modification.R"))
```

## Load in Data 

```{r load marvel object and significant gene tables}
# Load MARVEL object
setbp1_marvel <- read_rds(here::here(
  "data", "marvel",
  "setbp1_marvel_aligned.rds"
))

# Retrieve sample metadata
sample_metadata <- setbp1_marvel$sample.metadata

# Load significant gene tables
load(here::here("data", "marvel", "significant_tables.RData"))
```

## Make UpSet Plot

### Prep data for UpSet Plot with ComplexHeatmap

```{r make upset plot}
# make list of sig tables
tables_list <- mget(ls(pattern = "_sig_table"))

# remove all cell types from list
tables_list <- tables_list[2:8]

# pull out only genes
gene_lists <- map(tables_list, ~ select(.x, gene_short_name))

# make into character list
gene_lists <- map(gene_lists, ~ flatten(.))

# create matrix
gene_lists_mat <- ComplexHeatmap::list_to_matrix(gene_lists)

# rename columns
colnames(gene_lists_mat) <- c(
  "Astrocytes", "Excitatory Neurons",
  "Inhibitory Neurons", "Microglia",
  "Oligodendrocytes", "OPCs", "Vascular Cells"
)

# generate combination matrix
gene_lists_comb_mat <- make_comb_mat(gene_lists_mat)
```

### Draw and save UpSet Plot

```{r draw and save upsetplot}
# draw UpSet plot (ComplexHeatmap)
png(here::here("results", "upset_plots", "sig_gene_overlap.png"),
  width = 8, height = 5, units = "in", res = 300
)
gene_lists_upset_plot <- UpSet(gene_lists_comb_mat,
  comb_col = viridis(8)[2:7][comb_degree(gene_lists_comb_mat)],
  top_annotation = upset_top_annotation(gene_lists_comb_mat,
    add_numbers = TRUE, annotation_name_gp = gpar(fontface = "bold"),
    axis_param = list(gp = gpar(fontface = "bold")),
    numbers_gp = gpar(fontface = "bold")
  ),
  right_annotation = upset_right_annotation(gene_lists_comb_mat,
    add_numbers = TRUE, annotation_name_gp = gpar(fontface = "bold"),
    axis_param = list(gp = gpar(fontface = "bold")),
    numbers_gp = gpar(fontface = "bold")
  ),
  row_names_gp = gpar(fontface = "bold")
)
draw(gene_lists_upset_plot)
dev.off()

# view upsetplot
gene_lists_upset_plot
```

There is one gene that has signfificant PSI in all non-vascular cell types. This is the Son gene.

## Son gene analysis

#### Rename cell groups

```{r get cell groups}
# Pull cell types and matching ids
cell_type_ids <- sample_metadata %>%
  group_split(cell_type, .keep = TRUE) %>%
  map(~ set_names(.$cell.id, .$cell_type[1]))

# Rename
cell_group_list <- set_names(cell_type_ids, c(
  "Astrocytes", "Excitatory Neurons",
  "Inhibitory Neurons", "Microglia", "OPCs",
  "Oligodendrocytes", "Vascular Cells"
))
```

### Make MARVEL default plots for showing Son expression

#### MARVEL gene expression plot

```{r make marvel plot first}
setbp1_marvel <- adhocGene.TabulateExpression.Gene.10x(
  MarvelObject = setbp1_marvel,
  cell.group.list = cell_group_list,
  gene_short_name = "Son",
  min.pct.cells = 5,
  downsample = FALSE
)

setbp1_marvel$adhocGene$Expression$Gene$Plot
```

#### MARVEL SJ Expression Plot

```{r look at SJ usage}
setbp1_marvel <- adhocGene.TabulateExpression.PSI.10x(
  MarvelObject = setbp1_marvel,
  min.pct.cells = 5
)

setbp1_marvel$adhocGene$Expression$PSI$Plot
```

#### Plot gene and splice junction positions on genome track

This MARVEL function had errors, so I manually fixed its problems. It only shows one splice junction at a time. It also exports at a very low resolution, so I think it would be a better use of time to plot using some other package if we need it for publication.

```{r plot son gene plot}
setbp1_marvel <- adhocGene_PlotSJPosition_EJ(
  MarvelObject = setbp1_marvel,
  coord.intron = "chr16:91444873:91448517",
  rescale_introns = FALSE,
  show.protein.coding.only = TRUE,
  anno.label.size = 1.5
)

setbp1_marvel$adhocGene$SJPosition$Plot

# save
png(here::here("results", "marvel_outputs", "Son_SJ_1_location.png"),
  width = 12, height = 2, units = "in", res = 300
)
setbp1_marvel$adhocGene$SJPosition$Plot
dev.off()
```

## Calculate PSI Values

The current MARVEL setup doesn't have PSI values for every cell for every splice junction. we can remedy this ourselves.

Remember, PSI is the count of splice junctions divided by the counts for that gene multiplied by 100.

I also want to run this function on all splice junctions. Therefore we will create the required list of chromosome names.

```{r make list of chr names}
chr_prefixes <- paste0("chr", 1:19, ":")
chr_prefixes <- c(chr_prefixes, "chrX:", "chrY:")

writeLines(chr_prefixes, here::here("bin", "slurm", "chromosome_prefixes.txt"))
```

Now that the chromosome prefixes list has been saved, you may calculate psi for all chromosomes using script 04 (sbatch 04_submit_calc_psi.sh in terminal). For continuing Rmd analyses, move on to script 05 when that has finished running.

#### Style

```{r tidy script}
style_file("03_analyze_de_genes.Rmd")

lint("03_analyze_de_genes.Rmd",
  linters = linters_with_defaults(
    object_length_linter = NULL,
    object_name_linter = NULL,
    object_usage_linter = NULL
  )
)
```


#### Software versions

My software versions will be commented below.

```{r versions}
sessionInfo()
```
R version 4.3.1 (2023-06-16)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    grid      stats     graphics  grDevices utils     datasets  methods  
[9] base     

other attached packages:
 [1] gridExtra_2.3         data.table_1.14.8     Matrix_1.6-1.1       
 [4] wiggleplotr_1.24.0    GenomicRanges_1.52.1  GenomeInfoDb_1.36.4  
 [7] gtools_3.9.4          org.Mm.eg.db_3.17.0   org.Hs.eg.db_3.17.0  
[10] clusterProfiler_4.8.3 AnnotationDbi_1.62.2  IRanges_2.34.1       
[13] S4Vectors_0.38.2      Biobase_2.60.0        BiocGenerics_0.46.0  
[16] textclean_0.9.3       plyr_1.8.9            reshape2_1.4.4       
[19] ggrepel_0.9.4         ggnewscale_0.4.9      MARVEL_2.0.5         
[22] viridis_0.6.4         viridisLite_0.4.2     ComplexHeatmap_2.16.0
[25] patchwork_1.1.3       lubridate_1.9.3       forcats_1.0.0        
[28] stringr_1.5.1         dplyr_1.1.3           purrr_1.0.2          
[31] readr_2.1.4           tidyr_1.3.0           tibble_3.2.1         
[34] ggplot2_3.4.4         tidyverse_2.0.0       lintr_3.1.0          
[37] styler_1.10.2         here_1.0.1           

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3      jsonlite_1.8.7          rstudioapi_0.15.0      
  [4] shape_1.4.6             magrittr_2.0.3          farver_2.1.1           
  [7] rmarkdown_2.25          fs_1.6.3                GlobalOptions_0.1.2    
 [10] zlibbioc_1.46.0         vctrs_0.6.4             memoise_2.0.1          
 [13] RCurl_1.98-1.12         ggtree_3.8.2            htmltools_0.5.6.1      
 [16] qdapRegex_0.7.8         gridGraphics_0.5-1      desc_1.4.2             
 [19] cachem_1.0.8            igraph_1.5.1            lifecycle_1.0.4        
 [22] iterators_1.0.14        pkgconfig_2.0.3         gson_0.1.0             
 [25] R6_2.5.1                fastmap_1.1.1           GenomeInfoDbData_1.2.10
 [28] clue_0.3-65             aplot_0.2.2             digest_0.6.33          
 [31] enrichplot_1.20.3       colorspace_2.1-0        ps_1.7.5               
 [34] rprojroot_2.0.3         RSQLite_2.3.1           fansi_1.0.5            
 [37] timechange_0.2.0        polyclip_1.10-6         httr_1.4.7             
 [40] compiler_4.3.1          remotes_2.4.2.1         bit64_4.0.5            
 [43] withr_2.5.2             doParallel_1.0.17       downloader_0.4         
 [46] backports_1.4.1         BiocParallel_1.34.2     DBI_1.1.3              
 [49] ggforce_0.4.1           R.utils_2.12.2          MASS_7.3-60            
 [52] rjson_0.2.21            HDO.db_0.99.1           tools_4.3.1            
 [55] scatterpie_0.2.1        ape_5.7-1               R.oo_1.25.0            
 [58] glue_1.6.2              callr_3.7.3             nlme_3.1-163           
 [61] R.cache_0.16.0          GOSemSim_2.26.1         shadowtext_0.1.2       
 [64] cluster_2.1.4           fgsea_1.26.0            generics_0.1.3         
 [67] gtable_0.3.4            tzdb_0.4.0              R.methodsS3_1.8.2      
 [70] hms_1.1.3               tidygraph_1.2.3         xml2_1.3.5             
 [73] utf8_1.2.4              XVector_0.40.0          foreach_1.5.2          
 [76] pillar_1.9.0            yulab.utils_0.1.0       circlize_0.4.15        
 [79] splines_4.3.1           tweenr_2.0.2            treeio_1.24.3          
 [82] lattice_0.21-8          bit_4.0.5               tidyselect_1.2.0       
 [85] GO.db_3.17.0            Biostrings_2.68.1       knitr_1.44             
 [88] xfun_0.40               graphlayouts_1.0.1      matrixStats_1.0.0      
 [91] rex_1.2.1               stringi_1.8.1           ggfun_0.1.3            
 [94] lazyeval_0.2.2          yaml_2.3.7              evaluate_0.22          
 [97] codetools_0.2-19        ggraph_2.1.0            qvalue_2.32.0          
[100] ggplotify_0.1.2         cli_3.6.1               munsell_0.5.0          
[103] processx_3.8.2          xmlparsedata_1.0.5      Rcpp_1.0.11            
[106] png_0.1-8               parallel_4.3.1          blob_1.2.4             
[109] DOSE_3.26.2             bitops_1.0-7            tidytree_0.4.5         
[112] cyclocomp_1.1.1         scales_1.2.1            crayon_1.5.2           
[115] GetoptLong_1.0.5        rlang_1.1.2             cowplot_1.1.1          
[118] fastmatch_1.1-4         KEGGREST_1.40.1  
