---
title: "Combine PSI matrices"
output: html_document
date: "2024-02-09"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

The purpose of this script is to combine and visualize PSI matrices and splice junction information. It is dependent on running all seurat scripts, and marvel scripts 01 through 04. Script 04 is a bash and accompanying R script to submit parallel jobs for calculating PSI. Please use docker image setbp1_alternative_splicing:1.0.6.

## Load Packages

```{r load packages}
suppressPackageStartupMessages({
  # Load Lasseigne Lab standard packages
  library(here)
  library(styler)
  library(lintr)
  library(tidyverse)
  library(patchwork)
  library(ComplexHeatmap)
  library(viridis)

  # Load MARVEL package
  library(MARVEL)

  # Load adjunct MARVEL packages
  library(ggnewscale)
  library(ggrepel)
  library(reshape2)
  library(plyr)
  library(stringr)
  library(textclean)
  library(AnnotationDbi)
  library(clusterProfiler)
  library(org.Hs.eg.db)
  library(org.Mm.eg.db)
  library(gtools)
  library(GenomicRanges)
  library(IRanges)
  library(S4Vectors)
  library(wiggleplotr)
  library(Matrix)
  library(data.table)
  library(gridExtra)
})

# set seed
set.seed(123)

# source functions
source(here::here("src", "marvel", "functions.R"))
```

## Load in data

Lod in setbp1 marvel object

```{r load in data}
# Load MARVEL object
setbp1_marvel <- read_rds(here::here(
  "data", "marvel",
  "setbp1_marvel_aligned.rds"
))

# Retrieve sample metadata
sample_metadata <- setbp1_marvel$sample.metadata

# Save counts objects for easy accessibility
sj_counts <- setbp1_marvel[["sj.count.matrix"]]
gene_counts <- setbp1_marvel[["gene.count.matrix"]]

# Convert data type
sj_counts <- as(sj_counts, "CsparseMatrix")
gene_counts <- as(gene_counts, "CsparseMatrix")
```

These psi files are large, so this chunk takes a couple of minutes to run.

```{r pull in psi matrix data}
# List all files with the specified extension in the directory
files <- list.files(here::here("data", "marvel"),
  pattern = "_psi_matrix.Rds",
  full.names = TRUE
)

# Loop through each file and load it into the R environment
for (file in files) {
  # Extract the file name without extension
  file_name <- tools::file_path_sans_ext(basename(file))
  # Get object
  object <- read_rds(file)
  # Convert all NaNs and Infs to zero to reduce file size
  object@x[is.nan(object@x)] <- 0
  object@x[is.infinite(object@x)] <- 0
  # Any SJ with psi over 100 should be 100 maximum
  object@x[object@x > 100] <- 100
  # Drop non-structural zeroes
  object <- drop0(object)
  # Name the file
  assign(file_name, object)
}
```

## Combine data
Now, I must combine files.

```{r combine files into a list}
# List all objects in the environment
matching_objects <- ls(pattern = "_psi_matrix")
matching_objects <- matching_objects[1:21] # remove function

# Initialize an empty list to store the matched objects
matrices <- list()

# Loop through the matching objects and get them
for (object_name in matching_objects) {
  # Get the object from the environment
  object <- get(object_name)
  # Add the object to the list
  matrices[[object_name]] <- object
}
```

Now rbind each one together using do.call.

```{r combine matrices with do.call}
# rbind matrices together
full_psi_matrix <- do.call(rbind, matrices)

# rename the columns with cell ids
full_psi_matrix@Dimnames[[2]] <- sample_metadata$cell.id

# save the final psi matrix
write_rds(full_psi_matrix, here::here("data", "marvel", "psi_matrix.Rds"))
```

## Plot PSI distribution

```{r try plotting all nonzero values in x}
# save and plot nonzero psi values
png(here::here("results", "distributions", "psi_nonzero_hist.png"),
  width = 8, height = 6, units = "in", res = 300
)
hist(psi_matrix@x)
dev.off()
# save and plot nonzero log of sj counts
png(here::here("results", "distributions", "log_sj_counts_hist.png"),
  width = 8, height = 6, units = "in", res = 300
)
hist(log(setbp1_marvel$sj.count.matrix@x))
dev.off()
# save and plot nonzero normalized gene expression
png(here::here("results", "distributions", "gene_norm_hist.png"),
  width = 8, height = 6, units = "in", res = 300
)
hist(setbp1_marvel$gene.norm.matrix@x)
dev.off()
# save and plot nonzero log of gene expression
png(here::here("results", "distributions", "log_gene_counts_hist.png"),
  width = 8, height = 6, units = "in", res = 300
)
hist(log(setbp1_marvel$gene.count.matrix@x))
dev.off()
```

## Determine numbers of SJ detected per gene for each cell type

First, I'll split up the psi matrix by cell type

```{r get cell groups}
# Pull cell types and matching ids
cell_group_list <- sample_metadata %>%
  group_split(cell_type, .keep = TRUE) %>%
  map(~ set_names(.$cell.id, .$cell_type[1]))

# Rename
cell_group_list <- set_names(cell_group_list, c(
  "Astrocytes", "Excitatory Neurons",
  "Inhibitory Neurons", "Microglia", "OPCs",
  "Oligodendrocytes", "Vascular Cells"
))

# Pull mutant vs wildtype and matching ids
mutant_list <- sample_metadata %>%
  group_split(seq_folder, .keep = TRUE) %>%
  map(~ set_names(.$cell.id, .$seq_folder[[1]]))

# Rename
mutant_list <- set_names(mutant_list, c(
  "Mutant", "Wildtype"
))
```

Now split all matrices by cell type

```{r split matrices by cell type}
astrocytes_matrices <- subset_cell_type_matrices(cell_type = "Astrocytes")

excitatory_matrices <- subset_cell_type_matrices(
  cell_type = "Excitatory Neurons"
)

inhibitory_matrices <- subset_cell_type_matrices(
  cell_type = "Inhibitory Neurons"
)

microglia_matrices <- subset_cell_type_matrices(cell_type = "Microglia")

opcs_matrices <- subset_cell_type_matrices(cell_type = "OPCs")

oligodendrocyte_matrices <- subset_cell_type_matrices(
  cell_type = "Oligodendrocytes"
)

vascular_matrices <- subset_cell_type_matrices(cell_type = "Vascular Cells")
```


Now, we can make a dataframe with the total number of rows divided by number of cells

```{r get how many sjs per cell type}
sj_per_cell_type <- data.frame(
  "cell_type" = c(
    "Astrocytes", "Excitatory Neurons",
    "Inhibitory Neurons", "Microglia", "OPCs",
    "Oligodendrocytes", "Vascular Cells"
  ),
  "sj_number" = c(
    nrow(astrocytes_matrices[[2]]),
    nrow(excitatory_matrices[[2]]),
    nrow(inhibitory_matrices[[2]]),
    nrow(microglia_matrices[[2]]),
    nrow(oligodendrocyte_matrices[[2]]),
    nrow(opcs_matrices[[2]]),
    nrow(vascular_matrices[[2]])
  ),
  "sjs_per_gene" = c(
    nrow(astrocytes_matrices[[2]]) /
  nrow(astrocytes_matrices[[1]]),
nrow(excitatory_matrices[[2]]) /
  nrow(excitatory_matrices[[1]]),
nrow(inhibitory_matrices[[2]]) /
  nrow(inhibitory_matrices[[1]]),
nrow(microglia_matrices[[2]]) /
  nrow(microglia_matrices[[1]]),
nrow(oligodendrocyte_matrices[[2]]) /
  nrow(oligodendrocyte_matrices[[1]]),
nrow(opcs_matrices[[2]]) /
  nrow(opcs_matrices[[1]]),
nrow(vascular_matrices[[2]]) /
  nrow(vascular_matrices[[1]])
  )
)
```


```{r visualize splice junctions per cell type}

ggplot(sj_per_cell_type, aes(x = cell_type, y = sjs_per_gene, fill = cell_type)) +
  geom_bar(stat = "identity") +
  theme(legend.position = "none")

```

## Aggregate metrics by mutant and wildtype

I made a function called subset_mutant_matrices. I can use this to split the sj and gene count matrices

```{r split into mutant and wildtype}

astrocytes_mutants <- subset_mutant_matrices(astrocytes_matrices)

excitatory_mutants <- subset_mutant_matrices(excitatory_matrices)

inhibitory_mutants <- subset_mutant_matrices(inhibitory_matrices)

microglia_mutants <- subset_mutant_matrices(microglia_matrices)

oligodendrocyte_mutants <- subset_mutant_matrices(oligodendrocyte_matrices)

opcs_mutants <- subset_mutant_matrices(opcs_matrices)

vascular_mutants <- subset_mutant_matrices(vascular_matrices)

```

Beep boop

```{r make aggregated df for mutants}

sj_per_condition <- data.frame(
  "cell_type" = rep(c(
    "Astrocytes", "Excitatory Neurons",
    "Inhibitory Neurons", "Microglia", "OPCs",
    "Oligodendrocytes", "Vascular Cells"
  ), 2),
  
  "condition" = c(rep("Mutant", 7), rep("Wildtype", 7)),
  
  "sj_number" = c(
    nrow(astrocytes_mutants$Mutant[[2]]),
    nrow(excitatory_mutants$Mutant[[2]]),
    nrow(inhibitory_mutants$Mutant[[2]]),
    nrow(microglia_mutants$Mutant[[2]]),
    nrow(oligodendrocyte_mutants$Mutant[[2]]),
    nrow(opcs_mutants$Mutant[[2]]),
    nrow(vascular_mutants$Mutant[[2]]),
    nrow(astrocytes_mutants$Wildtype[[2]]),
    nrow(excitatory_mutants$Wildtype[[2]]),
    nrow(inhibitory_mutants$Wildtype[[2]]),
    nrow(microglia_mutants$Wildtype[[2]]),
    nrow(oligodendrocyte_mutants$Wildtype[[2]]),
    nrow(opcs_mutants$Wildtype[[2]]),
    nrow(vascular_mutants$Wildtype[[2]])
  ),
  
  "sjs_per_gene" = c(
    nrow(astrocytes_mutants$Mutant[[2]]) /
      nrow(astrocytes_mutants$Mutant[[1]]),
    nrow(excitatory_mutants$Mutant[[2]]) /
      nrow(excitatory_mutants$Mutant[[1]]),
    nrow(inhibitory_mutants$Mutant[[2]]) /
      nrow(inhibitory_mutants$Mutant[[1]]),
    nrow(microglia_mutants$Mutant[[2]]) /
      nrow(microglia_mutants$Mutant[[1]]),
    nrow(oligodendrocyte_mutants$Mutant[[2]]) /
      nrow(oligodendrocyte_mutants$Mutant[[1]]),
    nrow(opcs_mutants$Mutant[[2]]) /
      nrow(opcs_mutants$Mutant[[1]]),
    nrow(vascular_mutants$Mutant[[2]]) /
      nrow(vascular_mutants$Mutant[[1]]),
    nrow(astrocytes_mutants$Wildtype[[2]]) /
      nrow(astrocytes_mutants$Wildtype[[1]]),
    nrow(excitatory_mutants$Wildtype[[2]]) /
      nrow(excitatory_mutants$Wildtype[[1]]),
    nrow(inhibitory_mutants$Wildtype[[2]]) /
      nrow(inhibitory_mutants$Wildtype[[1]]),
    nrow(microglia_mutants$Wildtype[[2]]) /
      nrow(microglia_mutants$Wildtype[[1]]),
    nrow(oligodendrocyte_mutants$Wildtype[[2]]) /
      nrow(oligodendrocyte_mutants$Wildtype[[1]]),
    nrow(opcs_mutants$Wildtype[[2]]) /
      nrow(opcs_mutants$Wildtype[[1]]),
    nrow(vascular_mutants$Wildtype[[2]]) /
      nrow(vascular_mutants$Wildtype[[1]])
  ),
  
  "cell_count" = c(
  ncol(astrocytes_mutants$Mutant[[2]]),
  ncol(excitatory_mutants$Mutant[[2]]),
  ncol(inhibitory_mutants$Mutant[[2]]),
  ncol(microglia_mutants$Mutant[[2]]),
  ncol(oligodendrocyte_mutants$Mutant[[2]]),
  ncol(opcs_mutants$Mutant[[2]]),
  ncol(vascular_mutants$Mutant[[2]]),
  ncol(astrocytes_mutants$Wildtype[[2]]),
  ncol(excitatory_mutants$Wildtype[[2]]),
  ncol(inhibitory_mutants$Wildtype[[2]]),
  ncol(microglia_mutants$Wildtype[[2]]),
  ncol(oligodendrocyte_mutants$Wildtype[[2]]),
  ncol(opcs_mutants$Wildtype[[2]]),
  ncol(vascular_mutants$Wildtype[[2]])
)
)

```


```{r visualize splice junctions per cell type}
# plot
ggplot(sj_per_condition, aes(x = cell_type, y = sjs_per_gene, fill = condition)) +
  geom_bar(stat = "identity", position = position_dodge())

```

I notice there are more sjs per gene for all wts compared to controls, is this true for all cell counts?

```{r look at cell counts}
ggplot(sj_per_condition, aes(x = cell_type, y = cell_count, fill = condition)) +
  geom_bar(stat = "identity", position = position_dodge())
```
Yes, there are more cells so this score is imperfect and not something I want to draw major conclusions from.

#### Style

```{r tidy script}
style_file("05_join_psi_matrix.Rmd")

lint("05_join_psi_matrix.Rmd",
  linters = linters_with_defaults(
    object_length_linter = NULL,
    object_name_linter = NULL,
    object_usage_linter = NULL
  )
)
```


#### Software versions

My software versions will be commented below.

```{r versions}
sessionInfo()
```
R version 4.3.1 (2023-06-16)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    grid      stats     graphics  grDevices utils     datasets  methods  
[9] base     

other attached packages:
 [1] MatrixExtra_0.1.14    gridExtra_2.3         data.table_1.14.8    
 [4] Matrix_1.6-1.1        wiggleplotr_1.24.0    GenomicRanges_1.52.1 
 [7] GenomeInfoDb_1.36.4   gtools_3.9.4          org.Mm.eg.db_3.17.0  
[10] org.Hs.eg.db_3.17.0   clusterProfiler_4.8.3 AnnotationDbi_1.62.2 
[13] IRanges_2.34.1        S4Vectors_0.38.2      Biobase_2.60.0       
[16] BiocGenerics_0.46.0   textclean_0.9.3       plyr_1.8.9           
[19] reshape2_1.4.4        ggrepel_0.9.4         ggnewscale_0.4.9     
[22] MARVEL_2.0.5          viridis_0.6.4         viridisLite_0.4.2    
[25] ComplexHeatmap_2.16.0 patchwork_1.1.3       lubridate_1.9.3      
[28] forcats_1.0.0         stringr_1.5.1         dplyr_1.1.3          
[31] purrr_1.0.2           readr_2.1.4           tidyr_1.3.0          
[34] tibble_3.2.1          ggplot2_3.4.4         tidyverse_2.0.0      
[37] lintr_3.1.0           styler_1.10.2         here_1.0.1           

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3      jsonlite_1.8.7          rstudioapi_0.15.0      
  [4] shape_1.4.6             magrittr_2.0.3          farver_2.1.1           
  [7] rmarkdown_2.25          fs_1.6.3                GlobalOptions_0.1.2    
 [10] zlibbioc_1.46.0         vctrs_0.6.4             memoise_2.0.1          
 [13] RCurl_1.98-1.12         ggtree_3.8.2            htmltools_0.5.6.1      
 [16] qdapRegex_0.7.8         gridGraphics_0.5-1      desc_1.4.2             
 [19] cachem_1.0.8            igraph_1.5.1            lifecycle_1.0.4        
 [22] iterators_1.0.14        pkgconfig_2.0.3         gson_0.1.0             
 [25] R6_2.5.1                fastmap_1.1.1           GenomeInfoDbData_1.2.10
 [28] clue_0.3-65             aplot_0.2.2             digest_0.6.33          
 [31] enrichplot_1.20.3       colorspace_2.1-0        ps_1.7.5               
 [34] rprojroot_2.0.3         RSQLite_2.3.1           fansi_1.0.5            
 [37] timechange_0.2.0        polyclip_1.10-6         httr_1.4.7             
 [40] compiler_4.3.1          remotes_2.4.2.1         bit64_4.0.5            
 [43] withr_2.5.2             doParallel_1.0.17       downloader_0.4         
 [46] backports_1.4.1         BiocParallel_1.34.2     DBI_1.1.3              
 [49] ggforce_0.4.1           R.utils_2.12.2          float_0.3-2            
 [52] MASS_7.3-60             rjson_0.2.21            HDO.db_0.99.1          
 [55] tools_4.3.1             scatterpie_0.2.1        ape_5.7-1              
 [58] R.oo_1.25.0             glue_1.6.2              callr_3.7.3            
 [61] nlme_3.1-163            R.cache_0.16.0          GOSemSim_2.26.1        
 [64] shadowtext_0.1.2        cluster_2.1.4           fgsea_1.26.0           
 [67] generics_0.1.3          gtable_0.3.4            tzdb_0.4.0             
 [70] R.methodsS3_1.8.2       hms_1.1.3               tidygraph_1.2.3        
 [73] xml2_1.3.5              utf8_1.2.4              XVector_0.40.0         
 [76] foreach_1.5.2           pillar_1.9.0            yulab.utils_0.1.0      
 [79] circlize_0.4.15         splines_4.3.1           tweenr_2.0.2           
 [82] treeio_1.24.3           lattice_0.21-8          bit_4.0.5              
 [85] tidyselect_1.2.0        GO.db_3.17.0            Biostrings_2.68.1      
 [88] knitr_1.44              RhpcBLASctl_0.23-42     xfun_0.40              
 [91] graphlayouts_1.0.1      matrixStats_1.0.0       rex_1.2.1              
 [94] stringi_1.8.1           ggfun_0.1.3             lazyeval_0.2.2         
 [97] yaml_2.3.7              evaluate_0.22           codetools_0.2-19       
[100] ggraph_2.1.0            qvalue_2.32.0           ggplotify_0.1.2        
[103] cli_3.6.1               munsell_0.5.0           processx_3.8.2         
[106] xmlparsedata_1.0.5      Rcpp_1.0.11             png_0.1-8              
[109] parallel_4.3.1          blob_1.2.4              DOSE_3.26.2            
[112] bitops_1.0-7            tidytree_0.4.5          cyclocomp_1.1.1        
[115] scales_1.2.1            crayon_1.5.2            GetoptLong_1.0.5       
[118] rlang_1.1.2             cowplot_1.1.1           fastmatch_1.1-4        
[121] KEGGREST_1.40.1 
