---
title: "Import and Filter snRNA-seq Data"
author: "Emma Jones"
date: "2023-12-05"
output: html_document
---

The purpose of this script is to import and filter data using seurat. It is dependent on running the scripts in STARsolo_conda or downloading the data to start. Please use docker image setbp1_alternative_splicing:1.0.4 or above.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Loading Packages

```{r load packages}
suppressPackageStartupMessages({
  # Load Lasseigne Lab standard packages
  library(here)
  library(styler)
  library(lintr)

  # Load Seurat and other single-cell packages
  library(tidyverse)
  library(Seurat)
  library(patchwork)
  library(harmony)
  library(presto)
})

# set seed
set.seed(123)

# source functions
source(here("src", "seurat", "functions.R"))
```

## Load in Data

NOTE: SEURAT IS PICKY HERE, AND IF YOUR FILES ARE NOT GZIPPED IT WILL NOT READ THEM IN CORRECTLY. PLEASE MAKE SURE THESE FILES HAVE THE .GZ EXTENSION BEFORE PROCEEDING.

```{r load data}
# brain sample J1
J1_data <- Read10X(here(
  "data", "star", "J1", "Solo.out",
  "GeneFull_Ex50pAS", "filtered"
))

# brain sample J2
J2_data <- Read10X(here(
  "data", "star", "J2", "Solo.out",
  "GeneFull_Ex50pAS", "filtered"
))

# brain sample J3
J3_data <- Read10X(here(
  "data", "star", "J3", "Solo.out",
  "GeneFull_Ex50pAS", "filtered"
))

# brain sample J4
J4_data <- Read10X(here(
  "data", "star", "J4", "Solo.out",
  "GeneFull_Ex50pAS", "filtered"
))

# brain sample J13
J13_data <- Read10X(here(
  "data", "star", "J13", "Solo.out",
  "GeneFull_Ex50pAS", "filtered"
))

# brain sample J15
J15_data <- Read10X(here(
  "data", "star", "J15", "Solo.out",
  "GeneFull_Ex50pAS", "filtered"
))
```

Now that the data has been imported, we can create the Seurat object

```{r create Seurat object}
# brain sample 1 - S858R Cortex 1
brain_1 <- CreateSeuratObject(counts = J1_data, project = "setbp1")

brain_1[["sample_id"]] <- "J_1"

# brain sample 2 - WT Cortex 1
brain_2 <- CreateSeuratObject(counts = J2_data, project = "setbp1")

brain_2[["sample_id"]] <- "J_2"

# brain sample 3 - WT Cortex 2
brain_3 <- CreateSeuratObject(counts = J3_data, project = "setbp1")

brain_3[["sample_id"]] <- "J_3"

# brain sample 4 - WT Cortex 3
brain_4 <- CreateSeuratObject(counts = J4_data, project = "setbp1")

brain_4[["sample_id"]] <- "J_4"

# brain sample 13 - S858R Cortex 2
brain_13 <- CreateSeuratObject(counts = J13_data, project = "setbp1")

brain_13[["sample_id"]] <- "J_13"

# brain sample 15 - S858R Cortex 3
brain_15 <- CreateSeuratObject(counts = J15_data, project = "setbp1")

brain_15[["sample_id"]] <- "J_15"
```

Now we can merge these seurat objects

```{r merge seurat objects}
# merge S858R mutant samples
mutant_samples <- merge(brain_1,
  y = c(brain_13, brain_15),
  add.cell.ids = c("J1_het", "J13_het", "J15_het"),
  project = "setbp1"
)

mutant_samples[["condition"]] <- "mutant"

# merge wildtype samples
wt_samples <- merge(brain_2,
  y = c(brain_3, brain_4),
  add.cell.ids = c("J2_ctrl", "J3_ctrl", "J4_ctrl"),
  project = "setbp1"
)

wt_samples[["condition"]] <- "wildtype"


# merge all wildtype and mutant samples into one object
all_brain_samples <- merge(mutant_samples,
  y = wt_samples,
  project = "setbp1"
)

# view object
all_brain_samples
```

This is 56980 features across 53175 cells

## Quality Control

Now, we need to calculate and plot QC metrics. We can use Tabea Soelter's functions to expedite this process.

I'll start by getting the percent mitochondrial, but also run a modified version of QC from TS's functions.

```{r get percent_mt}
all_brain_samples[["percent_mt"]] <- PercentageFeatureSet(all_brain_samples,
  pattern = "^mt-"
)
```

I did a basic look at QC metrics as shown in the vignette, but we can also do our own QC.

```{r look at QC metrics}
VlnPlot(all_brain_samples,
  features = c("nFeature_RNA", "nCount_RNA", "percent_mt"),
  ncol = 3
)

plot1 <- FeatureScatter(all_brain_samples,
  feature1 = "nCount_RNA",
  feature2 = "percent_mt"
)

plot2 <- FeatureScatter(all_brain_samples,
  feature1 = "nCount_RNA",
  feature2 = "nFeature_RNA"
)

plot1 + plot2
```

Calculate more QC, specifically mitoRatio and log10GenesPerUMI

```{r calculate and plot more qc metrics}
# calculate qc
all_brain_samples <- calculate_qc(all_brain_samples)

# extra qc from metadata
metadata <- format_metadata(all_brain_samples)

# add this metadata back to seurat object
all_brain_samples@meta.data <- metadata

# plot results
pdf(here("results", "seurat_outputs", "qc_plots.pdf"))

plot_qc(metadata)

dev.off()
```

## Filtering

Now, we can filter our cells with after evaluating the QC metrics.

```{r subset data for downstream analysis}
# cell-level filtering
filtered_brain_samples <- subset(all_brain_samples,
  subset = nGene >= 1000 & nGene <= 15000 &
    mitoRatio < 0.05
)

# gene-level filtering, removing Malat1 bc it has extremely high expression
filtered_brain_samples <- filtered_brain_samples[!grepl(
  "Malat1",
  rownames(filtered_brain_samples)
), ]

# examine new filtered object
filtered_brain_samples
```

This now becomes 51465 cells.

Save object!

```{r save object}
# save
saveRDS(filtered_brain_samples,
  file = here("data", "seurat", "filtered_brain_samples.rds")
)
```

## Evaluate cell cycle effects and run PCA

```{r evaluate cell cycle effects and run PCA}
pdf(here("results", "seurat_outputs", "cellcycle_pca.pdf"))

filtered_brain_samples <- evaluate_cell_cycle(filtered_brain_samples)

dev.off()
```

Save object again!

```{r save object again}
# save
saveRDS(filtered_brain_samples,
  file = here("data", "seurat", "filtered_brain_samples_pca.rds")
)
```

## Harmony Integration

```{r harmony integration and run umap}
integrated_brain_samples <- RunHarmony(filtered_brain_samples,
  group.by.vars = "sample_id"
)

integrated_brain_samples <- RunUMAP(integrated_brain_samples,
  dims = 1:25,
  reduction = "harmony",
  reduction.name = "umap_harmony"
)
```

Harmony converged after 10 iterations

Save integrated object!

```{r save integrated object}
# save
saveRDS(integrated_brain_samples,
  file = here("data", "seurat", "integrated_brain_samples.rds")
)
```

## Clustering

```{r find clusters}
resolutions <- c(0.5, 0.75, 0.9, 1, 1.25)

clustered_brain_samples <- find_clusters(integrated_brain_samples,
  dims = 1:25,
  reduction = "harmony",
  resolutions = resolutions
)
```

Save clustered object!

```{r save clustered object}
# save
saveRDS(clustered_brain_samples,
  file = here("data", "seurat", "clustered_brain_samples.rds")
)
```

#### Style

```{r tidy script}
style_file("01_import_filter_data.Rmd")

lint("01_import_filter_data.Rmd",
  linters = linters_with_defaults(
    object_length_linter = NULL,
    object_name_linter = NULL,
    object_usage_linter = NULL
  )
)
```

#### Software versions

My software versions will be commented below.

```{r versions}
sessionInfo()
```
R version 4.3.1 (2023-06-16)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] presto_1.0.0       data.table_1.14.8  harmony_1.1.0      Rcpp_1.0.11       
 [5] patchwork_1.1.3    Seurat_5.0.0       SeuratObject_5.0.0 sp_2.1-1          
 [9] lubridate_1.9.3    forcats_1.0.0      stringr_1.5.1      dplyr_1.1.3       
[13] purrr_1.0.2        readr_2.1.4        tidyr_1.3.0        tibble_3.2.1      
[17] ggplot2_3.4.4      tidyverse_2.0.0    lintr_3.1.0        styler_1.10.2     
[21] here_1.0.1        

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3     rstudioapi_0.15.0      jsonlite_1.8.7        
  [4] magrittr_2.0.3         spatstat.utils_3.0-4   farver_2.1.1          
  [7] rmarkdown_2.25         vctrs_0.6.4            ROCR_1.0-11           
 [10] spatstat.explore_3.2-5 htmltools_0.5.6.1      sctransform_0.4.1     
 [13] parallelly_1.36.0      KernSmooth_2.23-22     htmlwidgets_1.6.2     
 [16] desc_1.4.2             ica_1.0-3              plyr_1.8.9            
 [19] plotly_4.10.3          zoo_1.8-12             igraph_1.5.1          
 [22] mime_0.12              lifecycle_1.0.4        pkgconfig_2.0.3       
 [25] Matrix_1.6-1.1         R6_2.5.1               fastmap_1.1.1         
 [28] fitdistrplus_1.1-11    future_1.33.0          shiny_1.7.5.1         
 [31] digest_0.6.33          colorspace_2.1-0       ps_1.7.5              
 [34] tensor_1.5             rprojroot_2.0.3        RSpectra_0.16-1       
 [37] irlba_2.3.5.1          labeling_0.4.3         progressr_0.14.0      
 [40] fansi_1.0.5            spatstat.sparse_3.0-3  timechange_0.2.0      
 [43] polyclip_1.10-6        abind_1.4-5            httr_1.4.7            
 [46] compiler_4.3.1         remotes_2.4.2.1        withr_2.5.2           
 [49] backports_1.4.1        fastDummies_1.7.3      R.utils_2.12.2        
 [52] MASS_7.3-60            tools_4.3.1            lmtest_0.9-40         
 [55] httpuv_1.6.12          future.apply_1.11.0    goftest_1.2-3         
 [58] R.oo_1.25.0            glue_1.6.2             callr_3.7.3           
 [61] nlme_3.1-163           R.cache_0.16.0         promises_1.2.1        
 [64] grid_4.3.1             Rtsne_0.16             cluster_2.1.4         
 [67] reshape2_1.4.4         generics_0.1.3         gtable_0.3.4          
 [70] spatstat.data_3.0-3    tzdb_0.4.0             R.methodsS3_1.8.2     
 [73] hms_1.1.3              xml2_1.3.5             utf8_1.2.4            
 [76] spatstat.geom_3.2-7    RcppAnnoy_0.0.21       ggrepel_0.9.4         
 [79] RANN_2.6.1             pillar_1.9.0           spam_2.10-0           
 [82] RcppHNSW_0.5.0         later_1.3.1            splines_4.3.1         
 [85] lattice_0.21-8         deldir_1.0-9           survival_3.5-7        
 [88] tidyselect_1.2.0       miniUI_0.1.1.1         pbapply_1.7-2         
 [91] knitr_1.44             gridExtra_2.3          scattermore_1.2       
 [94] xfun_0.40              matrixStats_1.0.0      rex_1.2.1             
 [97] stringi_1.8.1          lazyeval_0.2.2         yaml_2.3.7            
[100] evaluate_0.22          codetools_0.2-19       cli_3.6.1             
[103] uwot_0.1.16            xtable_1.8-4           reticulate_1.34.0     
[106] munsell_0.5.0          processx_3.8.2         xmlparsedata_1.0.5    
[109] spatstat.random_3.2-1  globals_0.16.2         png_0.1-8             
[112] parallel_4.3.1         ellipsis_0.3.2         dotCall64_1.1-0       
[115] listenv_0.9.0          viridisLite_0.4.2      cyclocomp_1.1.1       
[118] scales_1.2.1           ggridges_0.5.4         leiden_0.4.3          
[121] crayon_1.5.2           rlang_1.1.2            cowplot_1.1.1   