---
title: "Create Figure 5"
author: "Emma Jones"
date: "2024-05-13"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Create Figure 5

The point of this figure is to convey the importance of the Son gene and also add an additional panel to show where the splice junctions lit on transcript structures.

```{r load in packages}
suppressPackageStartupMessages({
  library(tidyverse)
  library(here)
  library(styler)
  library(lintr)
  library(cowplot)
  library(patchwork)
  library(viridis)
  library(MARVEL)
  library(ComplexHeatmap)
  library(Matrix)
  library(slam)
  library(circlize)
  library(ggtranscript)
  library(rtracklayer)
})

# source functions
source(here::here("src", "figures", "geom_split_violin.R"))
source(here::here("src", "figures", "functions.R"))
```

## Load data

```{r load in data}
# Load MARVEL data from marvel 04
load(here::here("data", "marvel", "marvel_04_split_counts.Rdata"))

# load in sj usage split by cell type
sj_usage_cell_type <- readRDS(here::here(
  "data", "marvel",
  "sj_usage_cell_type.rds"
))

# load in sj usage split by cell type and condition
sj_usage_condition <- readRDS(here::here(
  "data", "marvel",
  "sj_usage_condition.rds"
))

# set colors
cell_type_colors <- c(
  `Astrocytes` = "#6CA9E2",
  `Excitatory Neurons` = "#98D070",
  `Inhibitory Neurons` = "#DEE971",
  `Microglia` = "#B898E4",
  `Oligodendrocytes` = "#4AD8E6",
  `OPCs` = "#0A9A8D",
  `Vascular Cells` = "#E28C67"
)

annot_colors_list <- list(
  `Cell Type` = c(
    "Astrocytes" = "#6CA9E2",
    "Excitatory Neurons" = "#98D070",
    "Inhibitory Neurons" = "#DEE971",
    "Microglia" = "#B898E4",
    "Oligodendrocytes" = "#4AD8E6",
    "OPCs" = "#0A9A8D",
    "Vascular Cells" = "#E28C67"
  ),
  `Condition` = c("S858R+/-" = "#666666", "Wildtype" = "#C1C1C1")
)
```


```{r import gtf}
# set file path
gtf_file_path <- "/data/project/lasseigne_lab/GENOME_dir/GENCODE_mm39/release_M31/GTF/gencode.vM31.primary_assembly.annotation.gtf"

# import gtf
gtf <- rtracklayer::import(gtf_file_path)

# make into tibble
gtf <- gtf %>% dplyr::as_tibble()
```


```{r get splice junction expression}
# get normalized sj_expression for any further use
normalized_sj_expression <- read_rds(
  here::here("data", "marvel", "normalized_sj_expression.Rds")
)
```

## Define cell types

```{r get cell groups}
# Pull cell types and matching ids
cell_group_list <- sample_metadata %>%
  group_split(cell_type, .keep = TRUE) %>%
  map(~ set_names(.$cell.id, .$cell_type[1]))

# Rename
cell_types <- list()
for (i in seq_along(cell_group_list)) {
  cell_types[i] <- names(cell_group_list[[i]][1])
}

cell_group_list <- set_names(
  cell_group_list,
  cell_types
)

# Pull mutant vs wildtype and matching ids
mutant_list <- sample_metadata %>%
  group_split(seq_folder, .keep = TRUE) %>%
  map(~ set_names(.$cell.id, .$seq_folder[[1]]))

# Rename
condition_types <- list()
for (i in seq_along(mutant_list)) {
  condition_types[i] <- names(mutant_list[[i]][1])
}

mutant_list <- set_names(
  mutant_list,
  condition_types
)
```

## Data wrangle normalized expression

```{r pull out mutants and wildtype normalized expression}
mutant_ids <- mutant_list[["mutant"]]

# mutant
mutant_sj <-
  normalized_sj_expression[, colnames(normalized_sj_expression) == mutant_ids]

mutant_gene <-
  setbp1_marvel$gene.norm.matrix[
    , colnames(setbp1_marvel$gene.norm.matrix)
    == mutant_ids
  ]

# wildtype
wildtype_sj <-
  normalized_sj_expression[, !colnames(normalized_sj_expression) == mutant_ids]

wildtype_gene <-
  setbp1_marvel$gene.norm.matrix[
    , !colnames(setbp1_marvel$gene.norm.matrix)
    == mutant_ids
  ]
```
## Prepare data for plotting

```{r fix sj_usage_condition}
mutant_sju <- sj_usage_condition[1, ]
mutant_sju <- do.call(cbind, mutant_sju)
colnames(mutant_sju) <- names(cell_group_list)

wt_sju <- sj_usage_condition[2, ]
wt_sju <- do.call(cbind, wt_sju)
colnames(wt_sju) <- names(cell_group_list)
```

reformat cell type level data for visualization

```{r reformat cell type level data for visualization}
colnames(sj_usage_cell_type) <- names(cell_type_colors)

sj_usage_cell_type <- as.data.frame(sj_usage_cell_type)
sj_usage_cell_type$sj_position <- rownames(sj_usage_cell_type)
rownames(sj_usage_cell_type) <- NULL

sj_usage_cell_type <- pivot_longer(sj_usage_cell_type,
  cols = colnames(sj_usage_cell_type)[1:7], names_to = "cell_type",
  values_to = "sj_usage"
)

sj_usage_cell_type$sj_usage[is.infinite(sj_usage_cell_type$sj_usage)] <- NaN
sj_usage_cell_type$sj_usage[sj_usage_cell_type$sj_usage > 100] <- 100
```


```{r reformat mutant level data for visualization}
mutant_sju[is.infinite(mutant_sju)] <- NaN
mutant_sju[mutant_sju > 100] <- 100

colnames(mutant_sju) <- paste0(colnames(mutant_sju), "_mutant")
```


```{r reformat wt level data for visualization}
wt_sju[is.infinite(wt_sju)] <- NaN
wt_sju[wt_sju > 100] <- 100

colnames(wt_sju) <- paste0(colnames(wt_sju), "_wildtype")

# r bind together both conditions
conditions_sju <- cbind(mutant_sju, wt_sju)

# create delta sj usage matrix
delta_sju <- mutant_sju - wt_sju
```


```{r make annotation}
col_annot <- HeatmapAnnotation(
  `Cell Type` = names(cell_group_list),
  col = annot_colors_list, annotation_name_gp = gpar(fontface = "bold")
)
```


```{r reformat cell type level data for visualization}
# mutants
mutant_sj_expr_means <- list()
for (i in seq_along(cell_group_list)) {
  # get only mutant ids
  cell_group_mutants <- intersect(cell_group_list[[i]], mutant_list[["mutant"]])
  # split cell type
  cell_type_sj_expr <-
    normalized_sj_expression[
      ,
      colnames(normalized_sj_expression) %in% cell_group_mutants
    ]
  # get means
  mutant_sj_expr_means[[i]] <- rowMeans(cell_type_sj_expr)
}
names(mutant_sj_expr_means) <- names(cell_group_list)

mutant_sj_expr_means <- do.call(cbind, mutant_sj_expr_means)

mutant_sj_expr_means <- as.data.frame(mutant_sj_expr_means)

colnames(mutant_sj_expr_means) <-
  paste0(colnames(mutant_sj_expr_means), "_mutant")
```

### Get mean splice junction expression for wildtype and mutant cells

```{r get wt expr means}
# wildtypes
wildtype_sj_expr_means <- list()
for (i in seq_along(cell_group_list)) {
  # get only wildtype ids
  cell_group_wildtypes <-
    intersect(cell_group_list[[i]], mutant_list[["wildtype"]])
  # split cell type
  cell_type_sj_expr <-
    normalized_sj_expression[
      ,
      colnames(normalized_sj_expression) %in% cell_group_wildtypes
    ]
  # get means
  wildtype_sj_expr_means[[i]] <- rowMeans(cell_type_sj_expr)
}
names(wildtype_sj_expr_means) <- names(cell_group_list)

wildtype_sj_expr_means <- do.call(cbind, wildtype_sj_expr_means)

wildtype_sj_expr_means <- as.data.frame(wildtype_sj_expr_means)

colnames(wildtype_sj_expr_means) <-
  paste0(colnames(wildtype_sj_expr_means), "_wildtype")
```


```{r bind everything together!}
# now bind the 2 dfs
sj_expr_for_heatmap <- cbind(mutant_sj_expr_means, wildtype_sj_expr_means)

sj_expr_for_heatmap <- as.matrix(sj_expr_for_heatmap)

# make delta expression matrix
delta_expr_for_heatmap <-
  as.matrix(mutant_sj_expr_means) - as.matrix(wildtype_sj_expr_means)
```

```{r make dataframe for split violin}
base_df <- sample_metadata[, c("cell.id", "seq_folder", "cell_type")]

gene_expr_df <- make_norm_gene_expr_df("Son")
```

## Filter data for Son

```{r plot son again and tweak as needed}
# overall means
sj_expr_means <- rowMeans(normalized_sj_expression)

# get sjs
gene_sjs <-
  setbp1_marvel[["sj.metadata"]]$coord.intron[
    setbp1_marvel[["sj.metadata"]]$gene_short_name.start == "Son"
  ]

# get mean expression
sj_mean_expr <- sj_expr_means[gene_sjs]
```

```{r format gene sjs}
# get gene splice junctions
gene_junctions <- strsplit(gene_sjs, ":")

gene_junctions <- as_tibble(do.call(rbind, gene_junctions))

colnames(gene_junctions) <- c("seqnames", "start", "end")

gene_junctions$start <- as.numeric(gene_junctions$start)

gene_junctions$end <- as.numeric(gene_junctions$end)

gene_junctions$strand <- gtf$strand[gtf$gene_name == "Son"][1]

gene_junctions$mean_count <- sj_mean_expr

gene_junctions$long_name <- gene_sjs

gene_junctions <- gene_junctions[order(gene_junctions$long_name), ]
```

## Make split violin plot

```{r make split violin}
# make split violin plot of gene expression
split_violin <- ggplot(
  gene_expr_df,
  aes(
    x = cell_type, y = norm_expr,
    fill = cell_type, alpha = seq_folder
  )
) +
  geom_split_violin(show.legend = TRUE) +
  scale_alpha_manual(
    name = "Condition",
    labels = c("S858R+/-", "Wildtype"),
    values = c(1, 0.4),
    guide = guide_legend(
      override.aes =
        list(fill = "gray40", color = "black")
    )
  ) +
  theme_minimal(base_size = 12) +
  theme(
    axis.title = element_text(face = "bold"),
    axis.text = element_text(face = "bold", color = "black"),
    legend.text = element_text(face = "bold", size = 10),
    legend.title = element_text(face = "bold"),
    axis.text.x = element_text(angle = 45, hjust = 1)
  ) +
  scale_fill_manual(
    labels = names(cell_type_colors),
    values = cell_type_colors,
    guide = "none"
  ) +
  xlab("Cell Type") +
  ylab("Normalized\nGene Expression") +
  theme(legend.position = "bottom")
```

## Prepare data for heatmaps

```{r make heatmaps}
# get sj expr
sj_mean_expr <- sj_expr_means[gene_sjs]

# make splice junction expression heatmap
delta_sje_heatmap_df <-
  delta_expr_for_heatmap[rownames(delta_expr_for_heatmap) %in% gene_sjs, ]

print(gene_junctions$strand[1])

ifelse(gene_junctions$strand[1] == "+",
  rownames(delta_sje_heatmap_df) <-
    c(paste0("SJ-", seq_len(nrow(gene_junctions)))),
  rownames(delta_sje_heatmap_df) <-
    c(paste0("SJ-", rev(seq_len(nrow(gene_junctions)))))
)

# set colors
delta_gene_expr_cols <-
  colorRamp2(
    c(
      min(delta_sje_heatmap_df), 0,
      max(delta_sje_heatmap_df)
    ),
    c(RColorBrewer::brewer.pal(name = "BrBG", n = 3))
  )
```

## Create heatmap

```{r create heatmap}
# create heatmap
delta_gene_heatmap <- Heatmap(delta_sje_heatmap_df,
  name = "\u0394 Mean Norm\nSJ Expression",
  column_order = sort(colnames(delta_sje_heatmap_df)),
  col = delta_gene_expr_cols,
  top_annotation = col_annot, cluster_rows = FALSE,
  show_column_names = FALSE,
  row_title = "SJ Expression", border = TRUE,
  row_title_gp = gpar(fontface = "bold"),
  row_names_gp = gpar(
    fontface = "bold",
    col = c(rep("black", 2), rep("gray60", 19))
  )
)

# make splice junction usage heatmap
delta_sju_heatmap_df <-
  delta_sju[rownames(delta_sju) %in% gene_sjs, ]

ifelse(gene_junctions$strand[1] == "+",
  rownames(delta_sju_heatmap_df) <- c(
    paste0("SJ-", seq_len(nrow(gene_junctions)))
  ),
  rownames(delta_sju_heatmap_df) <- c(
    paste0("SJ-", rev(seq_len(nrow(gene_junctions))))
  )
)

# set colors
delta_sju_cols <-
  colorRamp2(
    c(
      min(delta_sju_heatmap_df), 0,
      max(delta_sju_heatmap_df)
    ),
    c(RColorBrewer::brewer.pal(name = "PRGn", n = 3))
  )

# make heatmap
delta_sj_heatmap <- Heatmap(delta_sju_heatmap_df,
  name = "\u0394 SJ Usage",
  column_order = sort(colnames(delta_sju_heatmap_df)),
  col = delta_sju_cols,
  cluster_rows = FALSE,
  show_column_names = FALSE,
  row_title = "SJ Usage", border = TRUE,
  row_title_gp = gpar(fontface = "bold"),
  row_names_gp = gpar(
    fontface = "bold",
    col = c(rep("black", 2), rep("gray60", 19))
  )
)

# combine heatmaps
both_delta_heatmaps <- delta_gene_heatmap %v% delta_sj_heatmap

# make heatmaps compatible with cowplot
delta_sj_expr_usage_heatmaps <- grid.grabExpr(draw(both_delta_heatmaps))
```

```{r make significance matrix}
son_signif_mtx <- matrix(
  data = NA, nrow = nrow(delta_sje_heatmap_df),
  ncol = ncol(delta_sje_heatmap_df)
)

colnames(son_signif_mtx) <- substr(
  colnames(delta_sje_heatmap_df), 1,
  nchar(colnames(delta_sje_heatmap_df)) - 7
)

rownames(son_signif_mtx) <- rownames(delta_sje_heatmap_df)

son_signif_mtx["SJ-1", ] <- c("**", NA, NA, NA, NA, "*", NA)

son_signif_mtx["SJ-2", ] <- c("**", "**", "**", "**", "**", "**", NA)
```

```{r add asterisks to heatmaps}
delta_gene_heatmap_sig <- Heatmap(delta_sje_heatmap_df,
  name = "\u0394 Mean Norm\nSJ Expression",
  column_order = sort(colnames(delta_sje_heatmap_df)),
  col = delta_gene_expr_cols,
  top_annotation = col_annot, cluster_rows = FALSE,
  show_column_names = FALSE,
  row_title = "SJ Expression", border = TRUE,
  row_title_gp = gpar(fontface = "bold"),
  row_names_gp = gpar(
    fontface = "bold",
    col = c(rep("black", 2), rep("gray60", 19))
  )
)

# make heatmap
delta_sj_heatmap_sig <- Heatmap(delta_sju_heatmap_df,
  name = "\u0394 SJ Usage",
  column_order = sort(colnames(delta_sju_heatmap_df)),
  col = delta_sju_cols,
  cluster_rows = FALSE,
  show_column_names = FALSE,
  row_title = "SJ Usage", border = TRUE,
  row_title_gp = gpar(fontface = "bold"),
  row_names_gp = gpar(
    fontface = "bold",
    col = c(rep("black", 2), rep("gray60", 19))
  ),
  cell_fun = function(j, i, x, y, w, h, fill) {
    if (!is.na(son_signif_mtx[i, j])) {
      grid.text(sprintf("%s", son_signif_mtx[i, j]), x, y,
        gp = gpar(fontsize = 10)
      )
    }
  }
)

# combine heatmaps
both_delta_heatmaps_sig <- delta_gene_heatmap_sig %v% delta_sj_heatmap_sig

both_delta_heatmaps_sig
# make heatmaps compatible with cowplot
delta_sj_expr_usage_heatmaps_sig <- grid.grabExpr(draw(both_delta_heatmaps_sig))
```

## Transcript visualization figure

```{r transcript visualization figure}
# filter annotation for specific gene
gene_annotation_from_gtf <- gtf %>%
  dplyr::filter(
    !is.na(gene_name),
    gene_name == "Son"
  )

gene_annotation_from_gtf <- gene_annotation_from_gtf %>%
  dplyr::select(
    seqnames,
    start,
    end,
    strand,
    type,
    gene_name,
    transcript_name,
    transcript_type
  )

# extract exons
gene_exons <- gene_annotation_from_gtf %>% dplyr::filter(type == "exon")

ifelse(gene_junctions$strand[1] == "+",
  gene_junctions$sj_label <-
    c(paste0("SJ-", seq_len(nrow(gene_junctions)))),
  gene_junctions$sj_label <-
    c(paste0("SJ-", rev(seq_len(nrow(gene_junctions)))))
)
```

```{r create actual figure}
# change colors and make pretty
gene_transcript_label <- gene_exons %>%
  ggplot(aes(
    xstart = start,
    xend = end,
    y = transcript_name
  )) +
  geom_range(
    aes(fill = transcript_type)
  ) +
  geom_intron(
    data = to_intron(gene_exons, "transcript_name"),
    aes(strand = strand)
  ) +
  theme_minimal() +
  theme(
    axis.title = element_text(face = "bold"),
    axis.text = element_text(face = "bold", color = "black"),
    legend.text = element_text(face = "bold"),
    legend.title = element_text(face = "bold"),
    axis.title.y = element_blank()
  ) +
  labs(linewidth = "Normalized SJ\nExpression", fill = "Transcript Type") +
  scale_fill_manual(
    labels = c(
      "NMD",
      "Protein Coding",
      "Protein Coding\nCDS Not Defined",
      "Retained Intron"
    ),
    values = c(viridis(6)[2:5])
  ) +
  xlab("Genomic Location")

gene_transcript_label <- gene_exons %>%
  ggplot(aes(
    xstart = start,
    xend = end,
    y = transcript_name
  )) +
  geom_range(
    aes(fill = transcript_type)
  ) +
  geom_intron(
    data = to_intron(gene_exons, "transcript_name"),
    aes(strand = strand)
  ) +
  theme_minimal() +
  theme(
    axis.title = element_text(face = "bold"),
    axis.text = element_text(face = "bold", color = "black"),
    legend.text = element_text(face = "bold"),
    legend.title = element_text(face = "bold"),
    axis.title.y = element_blank()
  ) +
  labs(linewidth = "Normalized SJ\nExpression", fill = "Transcript Type") +
  scale_fill_manual(
    labels = c(
      "NMD", "Protein Coding",
      "Protein Coding\nCDS Not Defined", "Retained Intron"
    ),
    values = c(viridis(6)[2:5])
  ) +
  xlab("Genomic Location") +
  theme(legend.position = "bottom") +
  guides(fill = guide_legend(nrow = 2, byrow = TRUE))

gene_transcript_label
```


```{r get son 202 cds and exons}
# obtain cds
son_cds <- gene_annotation_from_gtf %>% dplyr::filter(type == "CDS")

# extract exons and cds for the two transcripts to be compared
son_202_exons <- gene_exons %>% dplyr::filter(transcript_name == "Son-202")
son_202_cds <- son_cds %>% dplyr::filter(transcript_name == "Son-202")

# make dummy column as example
gene_junctions$sj_1_2_labs <- c("SJ-1", "SJ-2", rep(NA, 19))
```

show just son 202, one of the most "complete" transcripts

```{r son transcript structure}
son_202_plot <- son_202_exons %>%
  ggplot(aes(
    xstart = start,
    xend = end,
    y = "Son-202"
  )) +
  geom_range(
    fill = "white",
    height = 0.25
  ) +
  geom_range(
    data = son_202_cds,
    fill = viridis(6)[3]
  ) +
  geom_intron(
    data = to_intron(son_202_exons, "transcript_name")
  ) +
  geom_junction(
    data = gene_junctions, junction.y.max = 0.5,
    aes(linewidth = mean_count)
  ) +
  scale_linewidth(range = c(0.1, 1)) +
  geom_junction_label_repel(
    data = gene_junctions,
    aes(label = sj_label), junction.y.max = 0.5,
    segment.color = NA
  ) +
  theme_minimal() +
  theme(
    axis.title = element_text(face = "bold"),
    axis.text = element_text(face = "bold", color = "black"),
    legend.text = element_text(face = "bold"),
    legend.title = element_text(face = "bold"),
    axis.title.y = element_blank()
  ) +
  labs(linewidth = "Normalized SJ\nExpression") +
  xlab("Genomic Location") +
  theme(legend.position = "bottom")


options(ggrepel.max.overlaps = Inf)

son_202_plot
```

```{r zoom into 202 plot}
son_202_zoom <- son_202_plot + coord_cartesian(xlim = c(91445000, 91460000))

son_202_zoom
```

```{r arrange plots}
options(ggrepel.max.overlaps = Inf)

# arrange vln plot
gene_expr_vln <- plot_grid(son_202_zoom, split_violin,
  ncol = 1,
  labels = c("B", "C")
)

# arrange panels
paneled_figure <- plot_grid(gene_transcript_label, gene_expr_vln,
  delta_sj_expr_usage_heatmaps_sig,
  labels = c("A", "", "D"), nrow = 1
)

paneled_figure
```

### Save final figure 5

```{r save fig 5}
# save
png(here::here("results", "final_outputs", "figure5.png"),
  width = 14, height = 8, units = "in", res = 300
)
paneled_figure
dev.off()

# save son full transcript figure
png(here::here("results", "final_outputs", "supporting_fig_5.png"),
  width = 14, height = 3, units = "in", res = 300
)
son_202_plot
dev.off()
```

#### Save any extra info

```{r save csv}
write.csv(gene_junctions,
  here::here(
    "results", "supplementary_figures",
    "son_splice_junction_info.csv"
  ),
  row.names = FALSE
)
```

#### Style

```{r tidy script}
style_file("figure_5.Rmd")

lint("figure_5.Rmd",
  linters = linters_with_defaults(
    object_length_linter = NULL,
    object_name_linter = NULL,
    object_usage_linter = NULL
  )
)
```

#### Software versions

My software versions will be commented below.

```{r versions}
sessionInfo()
```
R version 4.3.1 (2023-06-16)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    grid      stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] rtracklayer_1.60.1    GenomicRanges_1.52.1  GenomeInfoDb_1.36.4   IRanges_2.34.1       
 [5] S4Vectors_0.38.2      BiocGenerics_0.46.0   ggtranscript_0.99.9   circlize_0.4.15      
 [9] slam_0.1-50           Matrix_1.6-1.1        ComplexHeatmap_2.16.0 MARVEL_2.0.5         
[13] viridis_0.6.4         viridisLite_0.4.2     patchwork_1.1.3       cowplot_1.1.1        
[17] lintr_3.1.0           styler_1.10.2         here_1.0.1            lubridate_1.9.3      
[21] forcats_1.0.0         stringr_1.5.1         dplyr_1.1.4           purrr_1.0.2          
[25] readr_2.1.4           tidyr_1.3.0           tibble_3.2.1          ggplot2_3.5.0        
[29] tidyverse_2.0.0      

loaded via a namespace (and not attached):
 [1] bitops_1.0-7                gridExtra_2.3               remotes_2.4.2.1            
 [4] rlang_1.1.3                 magrittr_2.0.3              clue_0.3-65                
 [7] GetoptLong_1.0.5            matrixStats_1.0.0           compiler_4.3.1             
[10] png_0.1-8                   callr_3.7.3                 vctrs_0.6.5                
[13] pkgconfig_2.0.3             shape_1.4.6                 crayon_1.5.2               
[16] backports_1.4.1             XVector_0.40.0              labeling_0.4.3             
[19] utf8_1.2.4                  Rsamtools_2.16.0            tzdb_0.4.0                 
[22] ps_1.7.5                    xfun_0.40                   zlibbioc_1.46.0            
[25] DelayedArray_0.26.7         BiocParallel_1.34.2         cluster_2.1.4              
[28] parallel_4.3.1              R6_2.5.1                    stringi_1.8.1              
[31] RColorBrewer_1.1-3          Rcpp_1.0.12                 SummarizedExperiment_1.30.2
[34] iterators_1.0.14            knitr_1.44                  R.utils_2.12.2             
[37] R.cache_0.16.0              timechange_0.2.0            tidyselect_1.2.1           
[40] rstudioapi_0.15.0           abind_1.4-5                 yaml_2.3.7                 
[43] doParallel_1.0.17           codetools_0.2-19            processx_3.8.2             
[46] lattice_0.21-8              plyr_1.8.9                  Biobase_2.60.0             
[49] withr_3.0.0                 desc_1.4.2                  xml2_1.3.5                 
[52] Biostrings_2.68.1           pillar_1.9.0                MatrixGenerics_1.12.3      
[55] foreach_1.5.2               rex_1.2.1                   generics_0.1.3             
[58] rprojroot_2.0.3             RCurl_1.98-1.12             hms_1.1.3                  
[61] munsell_0.5.1               scales_1.3.0                glue_1.7.0                 
[64] lazyeval_0.2.2              tools_4.3.1                 BiocIO_1.10.0              
[67] data.table_1.14.8           GenomicAlignments_1.36.0    XML_3.99-0.14              
[70] cyclocomp_1.1.1             colorspace_2.1-0            GenomeInfoDbData_1.2.10    
[73] restfulr_0.0.15             cli_3.6.2                   fansi_1.0.6                
[76] S4Arrays_1.0.6              gtable_0.3.4                R.methodsS3_1.8.2          
[79] digest_0.6.33               ggrepel_0.9.5               xmlparsedata_1.0.5         
[82] farver_2.1.1                rjson_0.2.21                R.oo_1.25.0                
[85] lifecycle_1.0.4             GlobalOptions_0.1.2   
