---
title: "Create Figure 2"
author: "Emma Jones"
date: "2024-02-21"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Create figure 2

The purpose of this script is to create a finalized version of figure 1. It is dependent on Marvel scripts 01 through 06. Run in docker 1.0.6.

## Load in packages

```{r load in packages}

suppressPackageStartupMessages({
  # Load Lasseigne Lab standard packages
  library(here)
  library(styler)
  library(lintr)
  library(tidyverse)
  library(patchwork)
  library(ComplexHeatmap)
  library(viridis)
  library(factoextra)

  # Load MARVEL package
  library(MARVEL)

  # Load adjunct MARVEL packages
  library(ggnewscale)
  library(ggrepel)
  library(reshape2)
  library(plyr)
  library(stringr)
  library(textclean)
  library(AnnotationDbi)
  library(clusterProfiler)
  library(org.Hs.eg.db)
  library(org.Mm.eg.db)
  library(gtools)
  library(GenomicRanges)
  library(IRanges)
  library(S4Vectors)
  library(wiggleplotr)
  library(Matrix)
  library(data.table)
  library(gridExtra)
})

# set seed
set.seed(123)

# source functions
source(here::here("src", "figures", "geom_split_violin.R"))

```

## Load in data

```{r load in data}
# load in PSI matrix from script 05
full_psi_matrix <- read_rds(here::here(
  "data", "marvel",
  "full_psi_matrix.Rds"
))

# Load MARVEL object
setbp1_marvel <- read_rds(here::here(
  "data", "marvel",
  "setbp1_marvel_aligned.rds"
))

# Retrieve sample metadata
sample_metadata <- setbp1_marvel$sample.metadata

# Retrieve sj metadata
sj_metadata <- setbp1_marvel$sj.metadata

# Save counts objects for easy accessibility
sj_counts <- setbp1_marvel[["sj.count.matrix"]]
gene_counts <- setbp1_marvel[["gene.count.matrix"]]

# Convert data type
sj_counts <- as(sj_counts, "CsparseMatrix")
gene_counts <- as(gene_counts, "CsparseMatrix")
```

## Format new metadata columns

```{r format metadata columns}

sample_metadata$num_sjs <- diff(sj_counts@p)
sample_metadata$num_genes <- diff(gene_counts@p)
sample_metadata$splicing_complexity <- diff(sj_counts@p)/diff(gene_counts@p)

```

## Set colors

```{r set colors}
cell_type_colors <- c(
  `Astrocytes` = "#6CA9E2",
  `Excitatory Neurons` = "#98D070",
  `Inhibitory Neurons` = "#DEE971",
  `Microglia` = "#B898E4",
  `Oligodendrocytes` = "#4AD8E6",
  `OPCs` = "#0A9A8D",
  `Vascular Cells` = "#E28C67"
)

condition_colors <- c(
  `mutant` = "#D8D8D8",
  `wildtype` = "#9E9E9E"
)

other_colors <- c("#0346a3", "#076466", "#2a6607")
```



## Make histogram panels

```{r make histrogram panels - psi}
# plot nonzero psi values
nonzero_psi <- full_psi_matrix@x

# make histogram with ggplot
nonzero_psi_histogram <- ggplot() + aes(nonzero_psi) +
  geom_histogram(fill = other_colors[2])  +
  theme_minimal(base_size = 15) +
  theme(
    axis.title = element_text(face = "bold"),
    axis.text = element_text(face = "bold", color = "black"),
    legend.text = element_text(face = "bold", size = 12),
    axis.text.x = element_text()
  )  +
  xlab("PSI (Non-zero)") +
  ylab("Count") 

nonzero_psi_histogram

```

```{r make histrogram panels - sj counts}
# plot nonzero log of sj counts
log_nonzero_sj_counts <- (log(setbp1_marvel$sj.count.matrix@x))

# make histogram with ggplot
log_nonzero_sj_counts_histogram <- ggplot() + aes(log_nonzero_sj_counts) +
  geom_histogram(fill = other_colors[1])  +
  theme_minimal(base_size = 15) +
  theme(
    axis.title = element_text(face = "bold"),
    axis.text = element_text(face = "bold", color = "black"),
    legend.text = element_text(face = "bold", size = 12),
    axis.text.x = element_text()
  )  +
  xlab("log(Splice Junction Counts)") +
  ylab("Count") 

log_nonzero_sj_counts_histogram

```

So, Seurat by default normalizes gene expression data by dividing by total counts for that cell, multiplied by a scale factor (typically 10,000), and natural-log transformed using log1p.

I wonder what it would look like to do that same scaling on the splice junction counts.

I'm guessing we may want to adjust the scale factor, since the row sums are in the hundreds, not thousands and tens of thousands. Therefore, I chose a scale factor of 1000.

```{r scale splice junction counts}
# copy sparse matrix
normalized_sj_expression <- sj_counts

# divide by column sums
normalized_sj_expression@x <- normalized_sj_expression@x / rep.int(colSums(normalized_sj_expression), diff(normalized_sj_expression@p))

# scale by 100
normalized_sj_expression@x <- normalized_sj_expression@x * 1000

# natural-log transform
normalized_sj_expression@x <- log1p(normalized_sj_expression@x)
```

Now that we normalized, we can try plotting again. There is also no need to log normalize since it is already normalized.

```{r plot normalized sj counts}
# plot quick histogram
hist(normalized_sj_expression@x)

# plot nice histograms

# make histogram with ggplot
nonzero_norm_sj_exp_histogram <- ggplot() + aes(normalized_sj_expression@x) +
  geom_histogram(fill = other_colors[1])  +
  theme_minimal(base_size = 15) +
  theme(
    axis.title = element_text(face = "bold"),
    axis.text = element_text(face = "bold", color = "black"),
    legend.text = element_text(face = "bold", size = 12),
    axis.text.x = element_text()
  )  +
  xlab("Normalized Splice Junction Expression (Non-zero)") +
  ylab("Count") 

nonzero_norm_sj_exp_histogram

```


```{r make histrogram panels}
# plot nonzero normalized gene expression

# make quick histogram
hist(setbp1_marvel$gene.norm.matrix@x)

# make histogram with ggplot
nonzero_norm_gene_exp_histogram <- ggplot() + aes(setbp1_marvel$gene.norm.matrix@x) +
  geom_histogram(fill = other_colors[3])  +
  theme_minimal(base_size = 15) +
  theme(
    axis.title = element_text(face = "bold"),
    axis.text = element_text(face = "bold", color = "black"),
    legend.text = element_text(face = "bold", size = 12),
    axis.text.x = element_text()
  )  +
  xlab("Normalized Gene Expression (Non-zero)") +
  ylab("Count") 

nonzero_norm_gene_exp_histogram

```

```{r plot splice junctions detected}

sj_num_split_violin <- ggplot(sample_metadata, aes(x = cell_type, y = num_sjs, fill = cell_type, alpha = seq_folder))  +
  geom_split_violin(show.legend = FALSE) + scale_alpha_manual(values = c(0.4, 1)) +
  theme_minimal(base_size = 15) +
  theme(
    axis.title = element_text(face = "bold"),
    axis.text = element_text(face = "bold", color = "black"),
    legend.text = element_text(face = "bold", size = 12),
    axis.text.x = element_text(angle = 45, hjust = 1)
  )  +
  scale_fill_manual(
    labels = names(cell_type_colors),
    values = cell_type_colors
  ) +
  xlab("Cell Type") +
  ylab("Splice Junctions Detected per Cell") 

sj_num_split_violin

```

```{r plot splicing complexity}

complexity_split_violin <- ggplot(sample_metadata, aes(x = cell_type, y = splicing_complexity, fill = cell_type, alpha = seq_folder))  +
  geom_split_violin(show.legend = FALSE) + scale_alpha_manual(values = c(0.4, 1)) +
  theme_minimal(base_size = 15) +
  theme(
    axis.title = element_text(face = "bold"),
    axis.text = element_text(face = "bold", color = "black"),
    legend.text = element_text(face = "bold", size = 12),
    axis.text.x = element_text(angle = 45, hjust = 1)
  )  +
  scale_fill_manual(
    labels = names(cell_type_colors),
    values = cell_type_colors
  ) +
  xlab("Cell Type") +
  ylab("Splicing Complexity") 

complexity_split_violin

```
## Assemble final figure

```{r make top row of histograms}

figure_2_top <- plot_grid(nonzero_norm_gene_exp_histogram, nonzero_norm_sj_exp_histogram, nonzero_psi_histogram, labels = c("A", "B", "C"), nrow = 1)

figure_2_top

figure_2_bottom <- plot_grid(sj_num_split_violin, complexity_split_violin, labels = c("D", "E"), nrow = 1)

figure_2_top

final_figure_2 <- plot_grid(figure_2_top, figure_2_bottom, ncol = 1, rel_heights = c(0.5,1))

final_figure_2

```

Now, we can save figure 2

```{r save figure 2}

png(here::here("results", "final_outputs", "figure2.png"),
  width = 12, height = 8, units = "in", res = 300
)
final_figure_2
dev.off()

```