---
title: "Create Figure 2"
author: "Emma Jones"
date: "2024-02-21"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Create figure 2 - ARCHIVED

The purpose of this script is to look at global splicing patterns, and what drives them in our data. We decided not to proceed with this analysis.

## Load in packages

```{r load in packages}
suppressPackageStartupMessages({
  # Load Lasseigne Lab standard packages
  library(here)
  library(styler)
  library(lintr)
  library(tidyverse)
  library(patchwork)
  library(ComplexHeatmap)
  library(viridis)
  library(factoextra)

  # Load MARVEL package
  library(MARVEL)

  # Load adjunct MARVEL packages
  library(ggnewscale)
  library(ggrepel)
  library(reshape2)
  library(plyr)
  library(stringr)
  library(textclean)
  library(AnnotationDbi)
  library(clusterProfiler)
  library(org.Hs.eg.db)
  library(org.Mm.eg.db)
  library(gtools)
  library(GenomicRanges)
  library(IRanges)
  library(S4Vectors)
  library(wiggleplotr)
  library(Matrix)
  library(data.table)
  library(gridExtra)
  library(cowplot)
  library(ggridges)
  library(ggpubr)
})

# set seed
set.seed(123)

# source functions
source(here::here("src", "figures", "geom_split_violin.R"))
```

## Load in data

```{r load in data}
# Load MARVEL data from marvel 04
load(here::here("data", "marvel", "marvel_04_split_counts.Rdata"))

# load normalized sj expression
normalized_sj_expression <- 
  load(here::here("data", "marvel", "normalized_sj_expression.Rdata"))
```


```{r get cell groups}
# Pull cell types and matching ids
cell_group_list <- sample_metadata %>%
  group_split(cell_type, .keep = TRUE) %>%
  map(~ set_names(.$cell.id, .$cell_type[1]))

# Rename
cell_group_list <- set_names(cell_group_list, c(
  "Astrocytes", "Excitatory Neurons",
  "Inhibitory Neurons", "Microglia", "OPCs",
  "Oligodendrocytes", "Vascular Cells"
))

# Pull mutant vs wildtype and matching ids
mutant_list <- sample_metadata %>%
  group_split(seq_folder, .keep = TRUE) %>%
  map(~ set_names(.$cell.id, .$seq_folder[[1]]))

# Rename
mutant_list <- set_names(mutant_list, c(
  "Mutant", "Wildtype"
))
```


## Split out sparse matrices

To use ggridges, I need to make a dataframe with a row for each value and a condition column that has mutant or wildtype.

```{r pull out mutants and wildtype normalized expression}
mutant_ids <- mutant_list[[1]]

# mutant
mutant_sj <-
  normalized_sj_expression[, colnames(normalized_sj_expression) == mutant_ids]

mutant_gene <-
  setbp1_marvel$gene.norm.matrix[
    , colnames(setbp1_marvel$gene.norm.matrix)
    == mutant_ids
  ]

# wildtype
wildtype_sj <-
  normalized_sj_expression[, !colnames(normalized_sj_expression) == mutant_ids]

wildtype_gene <-
  setbp1_marvel$gene.norm.matrix[
    , !colnames(setbp1_marvel$gene.norm.matrix)
    == mutant_ids
  ]
```

Need to now make dataframes which include each one of these.

Splice Junction Expression

```{r pull out nonzero values and make a dataframe - splice junction exp}

mutant_sj_vals <- data.frame("sj_expr" = mutant_sj@x, "Condition" = "Mutant")

wildtype_sj_vals <- data.frame("sj_expr" = wildtype_sj@x, "Condition" = "Wildtype")

sj_vals <- rbind(mutant_sj_vals, wildtype_sj_vals)

# test if distributions sig different

sj_result_ks <- ks.test(mutant_sj@x, wildtype_sj@x)

sj_result_ks
```


Gene expression

```{r pull out nonzero values and make a dataframe - gene expression}

mutant_gene_vals <- data.frame("gene_expr" = mutant_gene@x, "Condition" = "Mutant")

wildtype_gene_vals <- data.frame("gene_expr" = wildtype_gene@x, "Condition" = "Wildtype")

gene_vals <- rbind(mutant_gene_vals, wildtype_gene_vals)

# test if distributions sig different

gene_result_ks <- ks.test(mutant_gene@x, wildtype_gene@x)

gene_result_ks
```

## Set colors

```{r set colors}
cell_type_colors <- c(
  `Astrocytes` = "#6CA9E2",
  `Excitatory Neurons` = "#98D070",
  `Inhibitory Neurons` = "#DEE971",
  `Microglia` = "#B898E4",
  `Oligodendrocytes` = "#4AD8E6",
  `OPCs` = "#0A9A8D",
  `Vascular Cells` = "#E28C67"
)

condition_colors <- c(
  `mutant` = "#D8D8D8",
  `wildtype` = "#9E9E9E"
)

other_colors <- c("#0346a3", "#076466", "#2a6607")
```

First, I want to show the expression distributions across mutants and wildtypes.

## Make density plots

```{r format metadata columns}
sample_metadata$num_sjs <- diff(sj_counts@p)
sample_metadata$num_genes <- diff(gene_counts@p)
sample_metadata$num_sjs_genes <- diff(sj_counts@p) / diff(gene_counts@p)

mutant_ids <- sample_metadata$cell.id[sample_metadata$seq_folder == "mutant"]
wildtype_ids <- sample_metadata$cell.id[!sample_metadata$seq_folder == "mutant"]
```


### Splice junction expression

```{r make sj ridge plot}

nonzero_norm_sj_exp_ridge <-
  ggplot(sj_vals, aes(x = sj_expr, y = Condition, alpha = Condition)) + 
  geom_density_ridges(fill = other_colors[1], color = other_colors[1],
                      show.legend = FALSE)  +
  xlim(0, 6) +
  theme_minimal(base_size = 15) +
  scale_alpha_manual(values = c(0.7, 0.3)) +
  theme(
    axis.title = element_text(face = "bold"),
    axis.text = element_text(face = "bold", color = "black"),
    legend.text = element_text(face = "bold", size = 12),
    axis.text.x = element_text()
  ) +
  xlab("Normalized Splice Junction Expression") +
  ylab("Condition")

nonzero_norm_sj_exp_ridge
```
### Gene expression

```{r make sj ridge plot}

nonzero_norm_gene_exp_ridge <- 
  ggplot(gene_vals, aes(x = gene_expr, y = Condition, alpha = Condition)) + 
  geom_density_ridges(fill = other_colors[3], color = other_colors[3],
                      show.legend = FALSE)  +
  xlim(0, 6) +
  theme_minimal(base_size = 15) +
  scale_alpha_manual(values = c(0.7, 0.3)) +
  theme(
    axis.title = element_text(face = "bold"),
    axis.text = element_text(face = "bold", color = "black"),
    legend.text = element_text(face = "bold", size = 12),
    axis.text.x = element_text()
  ) +
  xlab("Normalized Gene Expression") +
  ylab("Condition")

nonzero_norm_gene_exp_ridge
```

## Make violin plots

### Splice junctions detected per cell

## Run linear regression, does number of genes explain number of sjs?

```{r run linear regression}

sjs_genes_lm <- lm(num_sjs ~ num_genes, data = sample_metadata)

summary(sjs_genes_lm)

```

Yes, number of genes can predict number of splice junctions per cell.

```{r make quick plot}

ggplot(sample_metadata, aes(x = num_genes, y = num_sjs)) +
  geom_point() +
  stat_poly_line() +
  stat_poly_eq(use_label(c("eq", "R2", "P"))) +
  scale_fill_manual(values = cell_type_colors) +
  scale_color_manual(values = cell_type_colors) +
  theme_minimal(base_size = 15) +
  theme(
    axis.title = element_text(face = "bold"),
    axis.text = element_text(face = "bold", color = "black"),
    legend.text = element_text(face = "bold", size = 12),
    legend.title = element_text(face = "bold"),
    axis.text.x = element_text()
  ) +
  guides(
    color = guide_legend(title = "Cell Type"),
    fill = guide_legend(title = "Cell Type")
  ) +
  xlab("Median Gene Expression") +
  ylab("Number of Splice Junctions")

```

## Average splice junctions / genes detected per cell

```{r compare means}

compare_means(num_sjs_genes ~ cell_type,  data = sample_metadata)

my_comparisions <- list(c(1,2),c(1,3), c(1,4), c(1,5), c(1,6), c(1,7), c(2,3),
                        c(2,4), c(2,5), c(2,6), c(2,7), c(3,4), c(3,5), c(3,6),
                        c(3,7), c(4,5), c(4,6), c(4,7), c(5,6), c(5,7), c(6,7))

```

### Combine both mutant and wildtype
```{r plot splice junctions normalized by number of genes expressed}
num_sj_gene_violin <- ggplot(
  sample_metadata,
  aes(
    x = cell_type, y = num_sjs_genes,
    fill = cell_type
  )
) +
  geom_violin(show.legend = FALSE) +
  stat_summary(fun.data = "mean_sdl", mult = 1, 
                 geom = "pointrange", show.legend = FALSE) +
  stat_compare_means(comparisons = my_comparisions, label = "p.signif") +
  theme_minimal(base_size = 15) +
  theme(
    axis.title = element_text(face = "bold"),
    axis.text = element_text(face = "bold", color = "black"),
    legend.text = element_text(face = "bold", size = 12),
    axis.text.x = element_text(angle = 45, hjust = 1)
  ) +
  scale_fill_manual(
    labels = names(cell_type_colors),
    values = cell_type_colors
  ) +
  xlab("Cell Type") +
  ylab("Splice Junctions / Genes")

num_sj_gene_violin
```

### Split Violin Plot

```{r plot splice junctions normalized by number of genes expressed}
num_sj_split_violin <- ggplot(
  sample_metadata,
  aes(
    x = cell_type, y = num_sjs,
    fill = cell_type, alpha = seq_folder
  )
) +
  geom_split_violin(show.legend = FALSE) +
  scale_alpha_manual(values = c(1, 0.4)) +
  theme_minimal(base_size = 15) +
  theme(
    axis.title = element_text(face = "bold"),
    axis.text = element_text(face = "bold", color = "black"),
    legend.text = element_text(face = "bold", size = 12),
    axis.text.x = element_text(angle = 45, hjust = 1)
  ) +
  scale_fill_manual(
    labels = names(cell_type_colors),
    values = cell_type_colors
  ) +
  xlab("Cell Type") +
  ylab("Splice Junctions")

num_sj_gene_split_violin


png(here::here("results", "marvel_outputs", "num_sj_split_violin.png"),
  width = 10, height = 8, units = "in", res = 300
)
num_sj_split_violin
dev.off()
```

```{r plot splice junctions normalized by number of genes expressed}
num_sj_gene_split_violin <- ggplot(
  sample_metadata,
  aes(
    x = cell_type, y = num_sjs_genes,
    fill = cell_type, alpha = seq_folder
  )
) +
  geom_split_violin(show.legend = FALSE) +
  scale_alpha_manual(values = c(1, 0.4)) +
  theme_minimal(base_size = 15) +
  theme(
    axis.title = element_text(face = "bold"),
    axis.text = element_text(face = "bold", color = "black"),
    legend.text = element_text(face = "bold", size = 12),
    axis.text.x = element_text(angle = 45, hjust = 1)
  ) +
  scale_fill_manual(
    labels = names(cell_type_colors),
    values = cell_type_colors
  ) +
  xlab("Cell Type") +
  ylab("Splice Junctions / Genes")

num_sj_gene_split_violin


png(here::here("results", "marvel_outputs", "num_sj_gene_split_violin.png"),
  width = 10, height = 8, units = "in", res = 300
)
num_sj_gene_split_violin
dev.off()
```
is num_sjs_genes normally distributed?

```{r is num_sjs_genes normally distributed?}

ks.test(sample_metadata$num_sjs_genes, pnorm)

```

No, at least not with this measure. so we want the wilcoxon rank-sum text then.
I could make a function to do this iteratively instead of doing a bunch of tests.

```{r wilcox between values}

mutant_metadata <- sample_metadata[sample_metadata$seq_folder == "mutant",]
wildtype_metadata <- sample_metadata[!sample_metadata$seq_folder == "mutant",]

cell_types <- unique(sample_metadata$cell_type)

wilcoxon_celltype_results <- lapply(cell_types, wilcoxon_test_celltype)

names(wilcoxon_celltype_results) <- cell_types

```

## Assemble final figure

```{r make top row of histograms}
figure_2_top <- plot_grid(nonzero_norm_gene_exp_ridge,
  nonzero_norm_sj_exp_ridge,
  labels = c("A", "B"), nrow = 1
)

figure_2_top

figure_2_bottom <- plot_grid(num_sj_gene_violin, num_sj_gene_split_violin, 
  labels = c("C", "D"), nrow = 1
)

figure_2_bottom

final_figure_2 <- plot_grid(figure_2_top, figure_2_bottom,
  ncol = 1,
  rel_heights = c(0.5, 1)
)

final_figure_2
```

Now, we can save figure 2

```{r save figure 2}
png(here::here("results", "final_outputs", "figure2.png"),
  width = 12, height = 8, units = "in", res = 300
)
final_figure_2
dev.off()
```

#### Style

```{r tidy script}
style_file("figure_2.Rmd")

lint("figure_2.Rmd",
  linters = linters_with_defaults(
    object_length_linter = NULL,
    object_name_linter = NULL,
    object_usage_linter = NULL
  )
)
```

#### Software versions

My software versions will be commented below.

```{r versions}
sessionInfo()
```
