---
title: "Psuedobulk DE Analysis"
output: html_document
date: "2024-03-19"
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


The purpose of this script is to run a pseudobulk gene expression analysis using the cell type data from seurat. It is dependent on seurat scripts 01 and 02. It should be run in docker 1.0.8.

Much of this analysis and code is based on the HBC training guide available here: https://hbctraining.github.io/scRNA-seq_online/lessons/pseudobulk_DESeq2_scrnaseq.html

## Load packages

```{r load packages}
suppressPackageStartupMessages({
  library(here)
  library(styler)
  library(lintr)
  library(tidyverse)
  library(patchwork)
  library(ComplexHeatmap)
  library(viridis)
  library(Seurat)
  library(Matrix.utils)
  library(edgeR)
  library(Matrix)
  library(reshape2)
  library(S4Vectors)
  library(SingleCellExperiment)
  library(pheatmap)
  library(apeglm)
  library(png)
  library(DESeq2)
  library(RColorBrewer)
  library(data.table)
})

source(here(
  "src",
  "deseq2",
  "functions.R"
))

# set seed
set.seed(123)
```

## Load data

```{r load annotated object}
seurat <-
  readRDS(file = here::here("data", "seurat", "annotated_brain_samples.rds"))
```

## Prepare data for DESeq2

```{r reformat seurat data to single cell experiment}
# extract raw counts and metadata
counts <- seurat@assays[["RNA"]]@layers[["counts"]]

# get gene names
rownames(counts) <- rownames(seurat)

metadata <- seurat@meta.data

# set up metadata as desired for aggregation and DE analysis
metadata$cluster_id <- factor(metadata$cell_type)

metadata$sample_id <- gsub("_", "", metadata$sample_id)
metadata$sample_id <- factor(metadata$sample_id)

# create single cell experiment object
sce <- SingleCellExperiment(
  assays = list(counts = counts),
  colData = metadata
)
```


```{r get cluster names}
# extract unique names of clusters
cluster_names <- levels(colData(sce)$cluster_id)
cluster_names

# get total number of clusters
length(cluster_names)
```


```{r get sample names}
# extract unique names of samples
sample_names <- levels(colData(sce)$sample_id)
sample_names

# get total number of samples
length(sample_names)
```


```{r get groups}
groups <- colData(sce)[, c("cluster_id", "sample_id")]
head(groups)
```


```{r aggregate expression matrix}
aggr_counts <- aggregate.Matrix(t(counts(sce)),
  groupings = groups, fun = "sum"
)

# explore output matrix
class(aggr_counts)
dim(aggr_counts)
aggr_counts[1:6, 1:6]
```


```{r transpose matrix}
aggr_counts <- t(aggr_counts)
aggr_counts[1:6, 1:6]
```

```{r extract counts for all cell types}
counts_ls <- list()

for (i in seq_along(cluster_names)) {
  ## Extract indexes of columns in the global matrix that match a given cluster
  column_idx <-
    which(tstrsplit(colnames(aggr_counts), "_")[[1]] == cluster_names[i])

  ## Store corresponding sub-matrix as one element of a list
  counts_ls[[i]] <- aggr_counts[, column_idx]
  names(counts_ls)[i] <- cluster_names[i]
}

# Explore the different components of the list
str(counts_ls)
```

```{r get necessary metadata}
# Extract sample-level variables
metadata <- colData(sce) %>%
  as.data.frame() %>%
  dplyr::select(seq_folder, sample_id)

# Exclude duplicated rows
metadata <- metadata[!duplicated(metadata), ]

dim(metadata)
head(metadata)

# restructure for fitting guide better
colnames(metadata) <- c("group_id", "sample_id")

# rename rows
rownames(metadata) <- metadata$sample_id
head(metadata)
```

```{r get cells per cell type for each sample}
t <- table(
  colData(sce)$sample_id,
  colData(sce)$cluster_id
)

t[1:6, 1:6]
```

```{r create metadata list}
## Initiate empty list
metadata_ls <- list()

for (i in seq_along(counts_ls)) {
  ## Initiate a data frame for cluster i
  df <- data.frame(cluster_sample_id = colnames(counts_ls[[i]]))

  ## Use tstrsplit() to separate cluster (cell type) and sample IDs
  df$cluster_id <- tstrsplit(df$cluster_sample_id, "_")[[1]]
  df$sample_id <- tstrsplit(df$cluster_sample_id, "_")[[2]]


  ## Retrieve cell count information for this cluster from cell count table
  idx <- which(colnames(t) == unique(df$cluster_id))
  cell_counts <- t[, idx]

  ## Remove samples with zero cell contributing to the cluster
  cell_counts <- cell_counts[cell_counts > 0]

  ## Match order of cell_counts and sample_ids
  sample_order <- match(df$sample_id, names(cell_counts))
  cell_counts <- cell_counts[sample_order]

  ## Append cell_counts to data frame
  df$cell_count <- cell_counts


  ## Join data frame
  df <- plyr::join(df, metadata,
    by = intersect(names(df), names(metadata))
  )

  ## Update rownames
  rownames(df) <- df$cluster_sample_id

  ## Store complete metadata for cluster i in list
  metadata_ls[[i]] <- df
  names(metadata_ls)[i] <- unique(df$cluster_id)
}

# Explore the different components of the list
str(metadata_ls)
```

## Perform differential gene expression using DESeq2

```{r choose cell type to begin}
idx <- which(names(counts_ls) == "Astrocytes")
cluster_counts <- counts_ls[[idx]]
cluster_metadata <- metadata_ls[[idx]]

cluster_metadata$group_id <- factor(cluster_metadata$group_id)

# check everything ok
all(colnames(cluster_counts) == rownames(cluster_metadata))
```



```{r create deseq2 object}
dds <- DESeqDataSetFromMatrix(cluster_counts,
  colData = cluster_metadata,
  design = ~group_id
)
```

Visualize data

```{r plot pca}
# Transform counts for data visualization
rld <- rlog(dds, blind = TRUE)

# Plot PCA
DESeq2::plotPCA(rld, ntop = 500, intgroup = "group_id")

DESeq2::plotPCA(rld, ntop = 500, intgroup = "cell_count")
# PC2 seems to influence cell count
```


```{r use hierarchical clustering}
# Extract the rlog matrix
rld_mat <- assay(rld)
rld_cor <- cor(rld_mat)

# Plot heatmap
pheatmap(rld_cor, annotation = cluster_metadata[, c("group_id"), drop = FALSE])
```

## Run DESeq2

```{r run deseq}
dds <- DESeq(dds)

# check dispersion estimates
plotDispEsts(dds)
```


```{r explore results}
# get results names
resultsNames(dds)

# generate results object
res <- results(dds, name = "group_id_wildtype_vs_mutant", alpha = 0.05)

# use apeglm
res <- lfcShrink(dds,
  coef = "group_id_wildtype_vs_mutant", res = res,
  type = "apeglm"
)
```


```{r look at results}
# Turn the DESeq2 results object into a tibble for use with tidyverse functions
res_tbl <- res %>%
  data.frame() %>%
  rownames_to_column(var = "gene") %>%
  as_tibble() %>%
  arrange(padj)

# Check results output
res_tbl

# Write all results to file
write.csv(res_tbl,
  here::here(
    paste0(
      "results/deseq2/",
      unique(cluster_metadata$cluster_id),
      "_",
      levels(cluster_metadata$group_id)[2],
      "_vs_",
      levels(cluster_metadata$group_id)[1],
      "_all_genes.csv"
    )
  ),
  quote = FALSE,
  row.names = FALSE
)
```

```{r filter for significance}
# Set thresholds
padj_cutoff <- 0.05

# Subset the significant results
sig_res <- dplyr::filter(res_tbl, padj < padj_cutoff) %>%
  dplyr::arrange(padj)

# Check significant genes output
sig_res

# Write significant results to file
write.csv(res_tbl,
  here::here(
    paste0(
      "results/deseq2/",
      unique(cluster_metadata$cluster_id),
      "_",
      levels(cluster_metadata$group_id)[2],
      "_vs_",
      levels(cluster_metadata$group_id)[1],
      "_signif_genes.csv"
    )
  ),
  quote = FALSE,
  row.names = FALSE
)
```


```{r set thresholds}
# Set thresholds
log2fc_cutoff <- 0.5

# Count significantly up/down genes above threshold
n_sig_up <- dplyr::filter(sig_res, log2FoldChange >= log2fc_cutoff) %>%
  nrow()
n_sig_dn <- dplyr::filter(sig_res, log2FoldChange <= -log2fc_cutoff) %>%
  nrow()
```




```{r scatterplot}
## Extract normalized counts from dds object
normalized_counts <- counts(dds, normalized = TRUE)
# I added below because there were no row names, but I wanted to index genes
# code: rownames(normalized_counts) <- 1:nrow(normalized_counts)


## Extract top 20 DEG from resLFC (make sure to order by padj)
top20_sig_genes <- sig_res %>%
  dplyr::arrange(padj) %>%
  dplyr::pull(gene) %>%
  head(n = 20)

## Extract matching normalized count values from matrix
top20_sig_counts <- normalized_counts[rownames(normalized_counts) %in%
                                        top20_sig_genes, ]
top20_sig_counts

## Convert wide matrix to long data frame for ggplot2
top20_sig_df <- data.frame(top20_sig_counts)
top20_sig_df$gene <- rownames(top20_sig_counts)

top20_sig_df <- melt(setDT(top20_sig_df),
  id.vars = c("gene"),
  variable.name = "cluster_sample_id"
) %>%
  data.frame()

## Replace "." by " " in cluster_sample_id variable (melt() introduced the ".")
top20_sig_df$cluster_sample_id <- gsub(
  "\\.",
  " ",
  top20_sig_df$cluster_sample_id
)
top20_sig_df

## Join counts data frame with metadata
top20_sig_df <- plyr::join(top20_sig_df, as.data.frame(colData(dds)),
  by = "cluster_sample_id"
)
top20_sig_df

## Generate plot
ggplot(top20_sig_df, aes(y = value, x = group_id, col = group_id)) +
  geom_jitter(height = 0, width = 0.15) +
  scale_y_continuous(trans = "log10") +
  ylab("log10 of normalized expression level") +
  xlab("condition") +
  ggtitle("Top 20 Significant DE Genes") +
  theme(plot.title = element_text(hjust = 0.5)) +
  facet_wrap(~gene)
```

```{r make heatmaps}
## Extract normalized counts for significant genes only
sig_counts <- normalized_counts[rownames(normalized_counts) %in% sig_res$gene, ]

## Set a color-blind friendly palette
heat_colors <- rev(brewer.pal(11, "PuOr"))

## Run pheatmap using the metadata data frame for the annotation
pheatmap(sig_counts,
  color = heat_colors,
  cluster_rows = TRUE,
  show_rownames = FALSE,
  annotation = cluster_metadata[, c("group_id", "cluster_id")],
  border_color = NA,
  fontsize = 10,
  scale = "row",
  fontsize_row = 10,
  height = 20
)
```

```{r volcano plot}
# Volcano plot
res_table_thres <- res_tbl[!is.na(res_tbl$padj), ] %>%
  mutate(threshold = padj < padj_cutoff & abs(log2FoldChange) >= log2fc_cutoff)
min(log10(res_table_thres$padj))

## Generate plot
ggplot(res_table_thres) +
  geom_point(aes(x = log2FoldChange, y = -log10(padj), colour = threshold)) +
  ggtitle("Volcano plot of S858R mutant astrocytes relative to control") +
  xlab("log2 fold change") +
  ylab("-log10 adjusted p-value") +
  scale_color_manual(values = c("grey60", "blue3")) +
  theme(
    legend.position = "none",
    plot.title = element_text(size = rel(1.3), hjust = 0.5),
    axis.title = element_text(size = rel(1.15))
  )
```
```{r run deseq2 on all cell types}
# run function on all clusters
map(
  cluster_names,
  get_dds_resultsAvsB,
  A = "wildtype",
  B = "mutant",
  save_path = "results/deseq2/"
)

# microglia and vascular cell do not have any significant genes, probably due to
# low cell counts
```




#### Style

```{r tidy script}
style_file("01_pseudobulk_analysis.Rmd")

lint("01_pseudobulk_analysis.Rmd",
  linters = linters_with_defaults(
    object_length_linter = NULL,
    object_name_linter = NULL,
    object_usage_linter = NULL
  )
)
```

#### Software versions

My software versions will be commented below.

```{r versions}
sessionInfo()
```
R version 4.3.1 (2023-06-16)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    grid      stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] data.table_1.14.8           RColorBrewer_1.1-3          DESeq2_1.40.2              
 [4] png_0.1-8                   apeglm_1.22.1               pheatmap_1.0.12            
 [7] SingleCellExperiment_1.22.0 SummarizedExperiment_1.30.2 Biobase_2.60.0             
[10] GenomicRanges_1.52.1        GenomeInfoDb_1.36.4         IRanges_2.34.1             
[13] MatrixGenerics_1.12.3       matrixStats_1.0.0           S4Vectors_0.38.2           
[16] BiocGenerics_0.46.0         reshape2_1.4.4              edgeR_3.42.4               
[19] limma_3.56.2                Matrix.utils_0.9.7          Matrix_1.6-1.1             
[22] Seurat_5.0.0                SeuratObject_5.0.0          sp_2.1-1                   
[25] viridis_0.6.4               viridisLite_0.4.2           ComplexHeatmap_2.16.0      
[28] patchwork_1.1.3             lubridate_1.9.3             forcats_1.0.0              
[31] stringr_1.5.1               dplyr_1.1.3                 purrr_1.0.2                
[34] readr_2.1.4                 tidyr_1.3.0                 tibble_3.2.1               
[37] ggplot2_3.4.4               tidyverse_2.0.0             lintr_3.1.0                
[40] styler_1.10.2               here_1.0.1                 

loaded via a namespace (and not attached):
  [1] RcppAnnoy_0.0.21        splines_4.3.1           later_1.3.1            
  [4] bitops_1.0-7            R.oo_1.25.0             polyclip_1.10-6        
  [7] rex_1.2.1               fastDummies_1.7.3       lifecycle_1.0.4        
 [10] doParallel_1.0.17       rprojroot_2.0.3         globals_0.16.2         
 [13] processx_3.8.2          lattice_0.21-8          MASS_7.3-60            
 [16] backports_1.4.1         magrittr_2.0.3          plotly_4.10.3          
 [19] rmarkdown_2.25          yaml_2.3.7              remotes_2.4.2.1        
 [22] httpuv_1.6.12           sctransform_0.4.1       spam_2.10-0            
 [25] spatstat.sparse_3.0-3   reticulate_1.34.0       cowplot_1.1.1          
 [28] pbapply_1.7-2           abind_1.4-5             zlibbioc_1.46.0        
 [31] Rtsne_0.16              R.cache_0.16.0          R.utils_2.12.2         
 [34] RCurl_1.98-1.12         circlize_0.4.15         GenomeInfoDbData_1.2.10
 [37] ggrepel_0.9.4           irlba_2.3.5.1           spatstat.utils_3.0-4   
 [40] listenv_0.9.0           goftest_1.2-3           RSpectra_0.16-1        
 [43] spatstat.random_3.2-1   fitdistrplus_1.1-11     parallelly_1.36.0      
 [46] leiden_0.4.3            codetools_0.2-19        DelayedArray_0.26.7    
 [49] xml2_1.3.5              tidyselect_1.2.0        shape_1.4.6            
 [52] farver_2.1.1            spatstat.explore_3.2-5  jsonlite_1.8.7         
 [55] GetoptLong_1.0.5        ellipsis_0.3.2          progressr_0.14.0       
 [58] ggridges_0.5.4          survival_3.5-7          iterators_1.0.14       
 [61] systemfonts_1.0.4       bbmle_1.0.25.1          foreach_1.5.2          
 [64] tools_4.3.1             ragg_1.2.5              ica_1.0-3              
 [67] Rcpp_1.0.11             glue_1.6.2              gridExtra_2.3          
 [70] xfun_0.40               numDeriv_2016.8-1.1     withr_2.5.2            
 [73] fastmap_1.1.1           fansi_1.0.5             callr_3.7.3            
 [76] digest_0.6.33           timechange_0.2.0        R6_2.5.1               
 [79] mime_0.12               textshaping_0.3.6       colorspace_2.1-0       
 [82] scattermore_1.2         tensor_1.5              spatstat.data_3.0-3    
 [85] R.methodsS3_1.8.2       utf8_1.2.4              generics_0.1.3         
 [88] httr_1.4.7              htmlwidgets_1.6.2       S4Arrays_1.0.6         
 [91] uwot_0.1.16             pkgconfig_2.0.3         gtable_0.3.4           
 [94] lmtest_0.9-40           XVector_0.40.0          htmltools_0.5.6.1      
 [97] dotCall64_1.1-0         clue_0.3-65             scales_1.2.1           
[100] cyclocomp_1.1.1         knitr_1.44              rstudioapi_0.15.0      
[103] tzdb_0.4.0              rjson_0.2.21            coda_0.19-4.1          
[106] nlme_3.1-163            bdsmatrix_1.3-7         zoo_1.8-12             
[109] GlobalOptions_0.1.2     KernSmooth_2.23-22      parallel_4.3.1         
[112] miniUI_0.1.1.1          desc_1.4.2              pillar_1.9.0           
[115] vctrs_0.6.4             RANN_2.6.1              promises_1.2.1         
[118] xtable_1.8-4            cluster_2.1.4           evaluate_0.22          
[121] mvtnorm_1.2-3           cli_3.6.1               locfit_1.5-9.8         
[124] compiler_4.3.1          rlang_1.1.2             crayon_1.5.2           
[127] grr_0.9.5               future.apply_1.11.0     labeling_0.4.3         
[130] emdbook_1.3.13          ps_1.7.5                plyr_1.8.9             
[133] stringi_1.8.1           deldir_1.0-9            BiocParallel_1.34.2    
[136] munsell_0.5.0           lazyeval_0.2.2          spatstat.geom_3.2-7    
[139] RcppHNSW_0.5.0          hms_1.1.3               future_1.33.0          
[142] shiny_1.7.5.1           ROCR_1.0-11             igraph_1.5.1           
[145] xmlparsedata_1.0.5
