---
title: "Import Setbp1 Data into Seurat"
author: "Emma Jones"
date: "2023-11-15"
output: html_document
---

The purpose of this script is to load in the nf-core scrnaseq pipeline results.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Loading Packages

```{r load packages}
suppressPackageStartupMessages({
  # Load Lasseigne Lab standard packages
  library(here)
  library(styler)
  library(lintr)

  # Load Seurat and other single-cell packages
  library(tidyverse)
  library(Seurat)
  library(patchwork)
  library(harmony)
  library(presto)
  })

# set seed
set.seed(123)

# source functions
source(here("src", "functions.R"))
```

## Load in Data

```{r load data}
# brain sample 1
brain_1_data <- Read10X(here("data", "nextflow", "star", "brain_1",
                             "brain_1.Solo.out", "Gene", "filtered"))

# brain sample 2
brain_2_data <- Read10X(here("data", "nextflow", "star", "brain_2",
                             "brain_2.Solo.out", "Gene", "filtered"))

# brain sample 3
brain_3_data <- Read10X(here("data", "nextflow", "star", "brain_3",
                             "brain_3.Solo.out", "Gene", "filtered"))

# brain sample 4
brain_4_data <- Read10X(here("data", "nextflow", "star", "brain_4",
                             "brain_4.Solo.out", "Gene", "filtered"))

# brain sample 5
brain_5_data <- Read10X(here("data", "nextflow", "star", "brain_5",
                             "brain_5.Solo.out", "Gene", "filtered"))

# brain sample 6
brain_6_data <- Read10X(here("data", "nextflow", "star", "brain_6",
                             "brain_6.Solo.out", "Gene", "filtered"))

```

Now that the data has been imported, we can create the seurat object

```{r create Seurat object}
# brain sample 1 - S858R Cortex 1
brain_1 <- CreateSeuratObject(counts = brain_1_data, project = "setbp1", min.cells = 3, min.features = 200)

brain_1[["sample_id"]] <- "S858R_1"

# brain sample 2 - WT Cortex 1
brain_2 <- CreateSeuratObject(counts = brain_2_data, project = "setbp1", min.cells = 3, min.features = 200)

brain_2[["sample_id"]] <- "WT_1"

# brain sample 3 - WT Cortex 2
brain_3 <- CreateSeuratObject(counts = brain_3_data, project = "setbp1", min.cells = 3, min.features = 200)

brain_3[["sample_id"]] <- "WT_2"

# brain sample 4 - WT Cortex 3
brain_4 <- CreateSeuratObject(counts = brain_4_data, project = "setbp1", min.cells = 3, min.features = 200)

brain_4[["sample_id"]] <- "WT_3"

# brain sample 5 - S858R Cortex 2
brain_5 <- CreateSeuratObject(counts = brain_5_data, project = "setbp1", min.cells = 3, min.features = 200)

brain_5[["sample_id"]] <- "S858R_2"

# brain sample 6 - S858R Cortex 3
brain_6 <- CreateSeuratObject(counts = brain_6_data, project = "setbp1", min.cells = 3, min.features = 200)

brain_6[["sample_id"]] <- "S858R_3"
```

Now we can merge these seurat objects

```{r merge seurat objects}
# merge wildtype samples
wt_samples <- merge(brain_2, y = c(brain_3, brain_4),
                    add.cell.ids = c("WT_1", "WT_2", "WT_3"),
                    project = "setbp1")

wt_samples[["condition"]] <- "wildtype"

# merge S858R mutant samples
mutant_samples <- merge(brain_1, y = c(brain_5, brain_6),
                    add.cell.ids = c("S858R_1", "S858R_2", "S858R_3"),
                    project = "setbp1")

mutant_samples[["condition"]] <- "mutant"

# merge all wildtype and mutant samples into one object
all_brain_samples <- merge(wt_samples,
                           y = mutant_samples,
                           project = "setbp1")

# view object
all_brain_samples
```

This is 31041 features across 45630 cells

## Quality Control

Now, we need to calculate and plot QC metrics. We can use Tabea Soelter's function to expedite this process.

I'll start by getting the percent mitochondrial, but also run a modified version of QC from TS's functions.

```{r get percent_mt}
all_brain_samples[["percent_mt"]] <- PercentageFeatureSet(all_brain_samples,
                                                          pattern = "^mt-")
```

I did a basic look at QC metrics as shown in the vignette, but we can also do our own QC.

```{r look at QC metrics}
VlnPlot(all_brain_samples,
        features = c("nFeature_RNA", "nCount_RNA", "percent_mt"),
        ncol = 3)

plot1 <- FeatureScatter(all_brain_samples, feature1 = "nCount_RNA", feature2 = "percent_mt")

plot2 <- FeatureScatter(all_brain_samples, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")

plot1 + plot2
```

Calculate more QC, specifically mitoRatio and log10GenesPerUMI

```{r calculate and plot more qc metrics}
# calculate qc
all_brain_samples <- calculate_qc(all_brain_samples)

# extra qc from metadata
metadata <- format_metadata(all_brain_samples)

# add this metadata back to seurat object
all_brain_samples@meta.data <- metadata

# plot results
pdf(here("results", "seurat_outputs", "qc_plots.pdf"))

plot_qc(metadata)

dev.off()
```

## Filtering

Now, we can filter our cells with after evaluating the QC metrics.

```{r subset data for downstream analysis}
# cell-level filtering
filtered_brain_samples <- subset(all_brain_samples,
                            subset = nGene > 500 & nGene < 7000 &
                              mitoRatio < 0.05 & log10GenesPerUMI > 0.85)

# examine new filtered object
filtered_brain_samples
```

This now becomes 45606 cells.

Save object!

```{r save object}
# save
saveRDS(filtered_brain_samples,
        file = here("data", "seurat", "filtered_brain_samples.rds"))
```

## Evaluate cell cycle effects and run PCA

```{r evaluate cell cycle effects and run PCA}

pdf(here("results", "seurat_outputs", "cellcycle_pca.pdf"))

filtered_brain_samples <- evaluate_cell_cycle(filtered_brain_samples)

dev.off()

```

Save object again!

```{r save object again}
# save
saveRDS(filtered_brain_samples,
        file = here("data", "seurat", "filtered_brain_samples.rds"))
```

## Harmony Integration

```{r harmony integration and run umap}

integrated_brain_samples <- RunHarmony(filtered_brain_samples,
                              group.by.vars = "sample_id")

integrated_brain_samples <- RunUMAP(integrated_brain_samples, dims = 1:25,
                                  reduction = "harmony",
                                  reduction.name = "umap_harmony")

```
Harmony converged after 9 iterations

Save integrated object!

```{r save integrated object}
# save
saveRDS(integrated_brain_samples,
        file = here("data", "seurat", "integrated_brain_samples.rds"))
```

## Clustering

```{r find clusters}

resolutions <- c(0.5, 0.75, 0.9, 1, 1.1, 1.2, 1.25, 1.5, 1.75)

clustered_brain_samples <- find_clusters(integrated_brain_samples, dims = 1:25,
                                          reduction = "harmony",
                                          resolutions = resolutions)

```

Save clustered object!

```{r save clustered object}
# save
saveRDS(clustered_brain_samples,
        file = here("data", "seurat", "clustered_brain_samples.rds"))
```


```{r find markers}
# set identity to best clustering after exploring data, I think 0.75 is enough
clustered_brain_samples <- SetIdent(clustered_brain_samples, value = "RNA_snn_res.0.75")

DimPlot(clustered_brain_samples)

marker_genes <- FindAllMarkers(clustered_brain_samples,
                               logfc.threshold = 0.2,
                               assay = "RNA",
                               min.pct = 0.01,
                               only.pos = TRUE)

top10 <- marker_genes %>%
  group_by(cluster) %>%
  top_n(-10, p_val_adj)
```


```{r featureplots}

FeaturePlot(clustered_brain_samples, features = c("Slc17a7"),
            label = TRUE) # excitatory neurons

FeaturePlot(clustered_brain_samples, features = c("Foxp2"),
            label = TRUE) # cluster in the middle

FeaturePlot(clustered_brain_samples, features = c("Elavl4"),
            label = TRUE)

FeaturePlot(clustered_brain_samples, features = c("Slc1a3", "Gja1"),
            label = TRUE)

FeaturePlot(clustered_brain_samples, features = c("Sst"),
            label = TRUE)
FeaturePlot(clustered_brain_samples, features = c("Pvalb"),
            label = TRUE)
FeaturePlot(clustered_brain_samples, features = c("Vip"),
            label = TRUE)
FeaturePlot(clustered_brain_samples, features = c("Lamp5"),
            label = TRUE)

FeaturePlot(clustered_brain_samples, features = c("Cx3cr1"),
            label = TRUE)
FeaturePlot(clustered_brain_samples, features = c("Csf1r"),
            label = TRUE)

FeaturePlot(clustered_brain_samples, features = c("Slc1a3"),
            label = TRUE)

FeaturePlot(clustered_brain_samples, features = c("Mbp"),
            label = TRUE) # oligodendrocytes
FeaturePlot(clustered_brain_samples, features = c("Pdgfra"),
            label = TRUE) # cluster 18 is OPCs?

FeaturePlot(clustered_brain_samples, features = c("Nr4a2"),
            label = TRUE) # I think 20 is fibroblasts

FeaturePlot(clustered_brain_samples, features = c("Gad1"),
            label = TRUE)

FeaturePlot(clustered_brain_samples, features = c("Satb2"))

FeaturePlot(clustered_brain_samples, features = c("Snap25"))
```

Look at cortical layer markers

```{r cortical layer markers}

# layer 2/3
FeaturePlot(clustered_brain_samples, features = c("Lamp5"))
# layer 2/4
FeaturePlot(clustered_brain_samples, features = c("Cux2"))
# layer 2/5
FeaturePlot(clustered_brain_samples, features = c("Rasgrf2"))
# layer 4/5
FeaturePlot(clustered_brain_samples, features = c("Rorb"))
# layer 5
FeaturePlot(clustered_brain_samples, features = c("Kcnk2", "Sulf2", "Pcp4", "Fezf2"))
# layer 5/6
FeaturePlot(clustered_brain_samples, features = c("Etv1", "Rprm", "Rxfp1", "Foxp2"))
# layer 6
FeaturePlot(clustered_brain_samples, features = c("Syt6", "Oprk1", "Nr4a2", "Synpr"))

```

Assign cell types

```{r assign cell types}

annotated_brain_samples <- RenameIdents(clustered_brain_samples,
                                        `1` = "Excitatory Neurons",
                                        `2` = "Excitatory Neurons",
                                        `3` = "Excitatory Neurons",
                                        `4` = "Excitatory Neurons",
                                        `5` = "Astrocytes",
                                        `6` = "Inhibitory Neurons",
                                        `7` = "Excitatory Neurons",
                                        `8` = "Excitatory Neurons",
                                        `9` = "Oligodendrocytes",
                                        `10` = "Excitatory Neurons",
                                        `11` = "Inhibitory Neurons",
                                        `12` = "Excitatory Neurons",
                                        `13` = "Inhibitory Neurons",
                                        `14` = "Excitatory Neurons",
                                        `15` = "Excitatory Neurons",
                                        `16` = "Excitatory Neurons",
                                        `17` = "Inhibitory Neurons",
                                        `18` = "OPCs",
                                        `19` = "Oligodendrocytes",
                                        `20` = "Fibroblasts",
                                        `21` = "Inhibitory Neurons",
                                        `22` = "Microglia",
                                        `23` = "Oligodendrocytes",
                                        `24` = "Astrocytes",
                                        `25` = "Excitatory Neurons",
                                        `26` = "Excitatory Neurons"
                                        )
```

Now that we have cell types identified, we can see what the final UMAP looks like.

```{r plot final umap}
png(here("results", "seurat_outputs", "annotated_umap.png"))
DimPlot(clustered_brain_samples)
dev.off()
```


Save annotated object!

```{r save annotated object}
# save
saveRDS(annotated_brain_samples,
        file = here("data", "seurat", "annotated_brain_samples.rds"))
```
